// @specifyapp/sdk/dist/builtInParsers/parsers-generation/to-css-font-import/convertTokens.js
import { isEqual, uniqWith } from 'lodash-es';
import { fontValueToCssFontImport } from './tokenTypes/font.js';
import { textStyleValueToCssFontImport } from './tokenTypes/textStyle.js';
import { specifyErrors } from '../../../errors/specifyErrors.js';
export const convertTokens = (tokens, { formats, fontsPath, includeFontWeight, genericFamily, fontDisplay }) => {
    const outputs = [];
    const warningMessages = [];
    for (const token of tokens) {
        try {
            const valueByMode = token.matchJSONValueByType({
                font: v => fontValueToCssFontImport(v, {
                    formats,
                    fontsPath,
                    includeFontWeight,
                    genericFamily,
                    fontDisplay,
                }),
                textStyle: v => textStyleValueToCssFontImport(v, {
                    formats,
                    fontsPath,
                    includeFontWeight,
                    genericFamily,
                    fontDisplay,
                }),
            }, () => undefined);
            if (!valueByMode)
                continue;
            const modes = Object.keys(valueByMode);
            for (const mode of modes) {
                outputs.push({ value: valueByMode[mode] });
            }
        }
        catch (error) {
            if (error.errorKey === 'SDTF_HAS_UNRESOLVABLE_ALIASES') {
                warningMessages.push({
                    type: 'warning',
                    content: `The token ${token.path.toString()} has unresolvable aliases.`,
                    errorKey: specifyErrors.PARSERS_ENGINE_PARSER_EXECUTION_FAILED.errorKey,
                });
            }
        }
    }
    return {
        outputs: uniqWith(outputs, isEqual),
        warningMessages,
    };
};
