// @specifyapp/sdk/dist/builtInParsers/internals/createBuiltInParserDefinition.js
import { z } from 'zod';
import { getParsersEngineDataBoxUnionSchema, } from '../../parsersEngine/definitions/parsersEngineDataBox.js';
import { commonBuiltInParserOptionsSchema, optionalCommonBuiltInParserOptionsSchema, } from './commonBuiltInParserOptions.js';
import { getParserOutputConfigurationUnionSchema, } from './parserOutputConfiguration.js';
export function createBuiltInParserDefinition(params) {
    const nameSchema = z.literal(params.name);
    const optionsSchema = params.optionsSchema
        ? params.hasOptionalOptions
            ? params.optionsSchema.merge(commonBuiltInParserOptionsSchema).optional()
            : params.optionsSchema.merge(commonBuiltInParserOptionsSchema)
        : optionalCommonBuiltInParserOptionsSchema;
    const outputTypes = params.outputTypes && params.outputTypes.length === 0 ? undefined : params.outputTypes;
    const outputSchema = outputTypes === undefined
        ? z.undefined().optional()
        : params.hasOptionalOutput
            ? getParserOutputConfigurationUnionSchema(outputTypes).optional()
            : getParserOutputConfigurationUnionSchema(outputTypes);
    return {
        name: params.name,
        nameSchema,
        kind: params.kind,
        hasOptionalOptions: params.hasOptionalOptions,
        optionsSchema: optionsSchema,
        hasOptionalOutput: params.hasOptionalOutput,
        outputTypes: outputTypes,
        outputSchema,
        inTypes: params.inTypes,
        inSchema: getParsersEngineDataBoxUnionSchema(params.inTypes),
        outType: params.outType,
        outSchema: getParsersEngineDataBoxUnionSchema([params.outType]),
        parserConfigurationSchema: z
            .object({
            name: nameSchema,
            options: optionsSchema,
            output: outputSchema,
        })
            .strict(),
    };
}
/* v8 ignore stop */
