// @specifyapp/specify-design-token-format/dist/mocks/tokenTypes/border.js
import { getMockedDimensionValue } from './dimension.js';
import { pickRandomInList, pickRandomNumberInRange } from '../../utils/mockGenerators.js';
import { getMockedBorderStyle } from './borderStyle.js';
import { getMockedColorValue } from './color.js';
export function getMockedBorderValue(partial, withNulls = true) {
    return {
        color: getMockedColorValue(),
        style: getMockedBorderStyle(),
        width: getMockedDimensionValue({
            value: pickRandomNumberInRange(1, 8),
            unit: pickRandomInList(['px', 'pt', 'em', 'rem']),
        }),
        rectangleCornerRadii: withNulls
            ? null
            : [
                getMockedDimensionValue({
                    value: pickRandomNumberInRange(1, 8),
                    unit: pickRandomInList(['px', 'pt', 'em', 'rem']),
                }),
                getMockedDimensionValue({
                    value: pickRandomNumberInRange(1, 8),
                    unit: pickRandomInList(['px', 'pt', 'em', 'rem']),
                }),
                getMockedDimensionValue({
                    value: pickRandomNumberInRange(1, 8),
                    unit: pickRandomInList(['px', 'pt', 'em', 'rem']),
                }),
                getMockedDimensionValue({
                    value: pickRandomNumberInRange(1, 8),
                    unit: pickRandomInList(['px', 'pt', 'em', 'rem']),
                }),
            ],
        ...partial,
    };
}
