// @specifyapp/sdk/src/builtInParsers/shared/to-scss-mixin-text-style/tokenTypes/textStyle.ts
import { SpecifyTextStyleValue } from '@specifyapp/specify-design-token-format';
import { PossibleTextStylesKeysSchemaType } from '../../definition.js';
import { colorValueToScssMixin } from './color.js';
import { dimensionValueToScssMixin } from './dimension.js';
import { fontWeightValueToScssMixin } from './fontWeight.js';

const allPossibleKeys: PossibleTextStylesKeysSchemaType = [
  'fontFamily',
  'fontWeight',
  'fontSize',
  'color',
  'fontFeatures',
  'lineHeight',
  'letterSpacing',
  'paragraphSpacing',
  'textAlignHorizontal',
  'textAlignVertical',
  'textDecoration',
  'textIndent',
  'textTransform',
];

const keysInCss = [
  'color',
  'font-family',
  'font-size',
  'line-height',
  'letter-spacing',
  'text-align',
  'vertical-align',
  'text-transform',
  'font-variant',
  'text-decoration',
  'text-indent',
  'font-weight',
  'font-style',
] as const;

function createFinalKeysToKeepList(
  exclude: PossibleTextStylesKeysSchemaType,
  include: PossibleTextStylesKeysSchemaType,
) {
  if (exclude.length > 0) {
    return allPossibleKeys.filter(key => !exclude.includes(key));
  }

  if (include.length > 0) {
    return allPossibleKeys.filter(key => include.includes(key));
  }

  return allPossibleKeys;
}

export function textStyleValueToScssMixin(
  textStyle: SpecifyTextStyleValue,
  {
    exclude,
    include,
    genericFamily,
  }: {
    exclude: PossibleTextStylesKeysSchemaType;
    include: PossibleTextStylesKeysSchemaType;
    genericFamily: string;
  },
) {
  const value = {} as Record<typeof keysInCss[number], string | number>;
  const keysToAdd = createFinalKeysToKeepList(exclude, include);
  // color
  const color = textStyle.color ? colorValueToScssMixin(textStyle.color) : null;
  if (color && keysToAdd.includes('color')) {
    value['color'] = color;
  }
  // font-family
  // Use family here because of how we convert tokens from token studio
  const fontFamily = textStyle.font.family;
  if (keysToAdd.includes('fontFamily')) {
    value['font-family'] = `${fontFamily}${genericFamily ? `, ${genericFamily}` : ''}`;
  }
  // font-size
  const fontSize = dimensionValueToScssMixin(textStyle.fontSize);
  if (fontSize && keysToAdd.includes('fontSize')) {
    value['font-size'] = fontSize;
  }
  // line-height

  if (textStyle.lineHeight && keysToAdd.includes('lineHeight')) {
    const lineHeightToUse = textStyle.lineHeight;
    value['line-height'] = dimensionValueToScssMixin(lineHeightToUse);
  }
  // letter-spacing
  const letterSpacing = textStyle.letterSpacing
    ? dimensionValueToScssMixin(textStyle.letterSpacing)
    : null;
  // Don't want it if letter spacing = 0
  if (textStyle.letterSpacing?.value && letterSpacing && keysToAdd.includes('letterSpacing')) {
    value['letter-spacing'] = letterSpacing;
  }
  // text-align
  const textAlign = textStyle.textAlignHorizontal;
  if (textAlign && keysToAdd.includes('textAlignHorizontal')) {
    value['text-align'] = textAlign;
  }
  // vertical-align
  const verticalAlign = textStyle.textAlignVertical;
  if (verticalAlign && keysToAdd.includes('textAlignVertical')) {
    value['vertical-align'] = verticalAlign;
  }
  // text-transform
  const textTransform = textStyle.textTransform;
  // Don't want it if text transform is none
  if (textTransform && textTransform !== 'none' && keysToAdd.includes('textTransform')) {
    value['text-transform'] = textTransform;
  }
  // font-variant
  const fontVariant = textStyle.fontFeatures?.join(' ');
  // Don't want it if font variant is none
  if (fontVariant && fontVariant !== 'none' && keysToAdd.includes('fontFeatures')) {
    value['font-variant'] = fontVariant;
  }
  // text-decoration
  const textDecoration = textStyle.textDecoration;
  if (textDecoration && textDecoration !== 'none' && keysToAdd.includes('textDecoration')) {
    value['text-decoration'] = textDecoration;
  }
  // text-indent
  const textIndent = textStyle.textIndent ? dimensionValueToScssMixin(textStyle.textIndent) : null;
  // Don't want it if text indent is 0
  if (textStyle.textIndent?.value && textIndent && keysToAdd.includes('textIndent')) {
    value['text-indent'] = textIndent;
  }
  // font-weight
  const fontWeight = fontWeightValueToScssMixin(textStyle.font.weight);
  if (fontWeight && keysToAdd.includes('fontWeight')) {
    value['font-weight'] = fontWeight;
  }
  // font-style
  const fontStyle = textStyle.font.style;
  if (fontStyle === 'italic') {
    value['font-style'] = fontStyle;
  }

  return value;
}
