// @specifyapp/sdk/src/httpApi/routes/executeParsersEngineRPC.ts
import { z } from 'zod';
import { serializedParsersEngineDataBoxSchema } from '../../parsersEngine/definitions/parsersEngineDataBox.js';
import { SerializedParsersEngineResults } from '../../parsersEngine/index.js';

export function makeExecuteParsersEngineRPCRoute() {
  return `/v2/parsers-engine-rpc` as const;
}

export const executeParsersEngineRPCRouteRequestBodySchema = z
  .object({
    dataBox: serializedParsersEngineDataBoxSchema,
    rules: z.array(z.record(z.string(), z.unknown())),
    returnedKeys: z
      .object({
        output: z.boolean().optional(),
        next: z.boolean().optional(),
        errorMessages: z.boolean().optional(),
        warningMessages: z.boolean().optional(),
        informationMessages: z.boolean().optional(),
      })
      .strict()
      .optional(),
  })
  .strict();

export type ExecuteParsersEnginRPCRouteRequestBody = z.infer<
  typeof executeParsersEngineRPCRouteRequestBodySchema
>;

export type ExecuteParsersEngineRPCRouteResponseBody = SerializedParsersEngineResults;
