// @specifyapp/sdk/src/httpApi/routes/getRepositoryTokenTreeByOwnerAndName.ts
import { z } from 'zod';
import { SpecifyDesignTokenFormat } from '@specifyapp/specify-design-token-format';

import { createPublicHttpRouteDefinition } from '../internals/createPublicHttpRouteDefinition.js';

export function makeGetRepositoryTokenTreeByOwnerAndNameRoute(options?: {
  owner: string;
  name: string;
}) {
  const { owner, name } = options ?? {};
  return `/v2/${owner ?? '{owner}'}/repository/${name ?? '{name}'}/tokens` as const;
}

const getRepositoryTokenTreeByOwnerAndNameParamsSchema = z
  .object({
    owner: z.string(),
    name: z.string(),
  })
  .strict();

export const getRepositoryTokenTreeByOwnerAndNameRouteDefinition = createPublicHttpRouteDefinition({
  authenticationStrategy: 'personalAccessToken',
  httpMethod: 'GET',
  makeRoute: makeGetRepositoryTokenTreeByOwnerAndNameRoute,
  responsePayloadSchema: z
    .object({
      id: z.string(),
      tokenTree: z.custom<SpecifyDesignTokenFormat>(v => v),
    })
    .strict(),
  paramsSchema: getRepositoryTokenTreeByOwnerAndNameParamsSchema,
});
