// @specifyapp/sdk/src/builtInParsers/shared/to-javascript/tokenToAstToken.ts
import { SpecifyDesignTokenTypeName, TokenState } from '@specifyapp/specify-design-token-format';
import { jsonValueMatcher, matchJsonValue } from '../../utils/jsonValueMatcher.js';

export type AstMetadataPaths = { [type in SpecifyDesignTokenTypeName]?: Set<string> };

export type AstMetadata = {
  types?: Set<SpecifyDesignTokenTypeName>;
  paths: Set<string>;
  modes: Set<string>;
};

export type ValueWithMode = { [mode: string]: Value };

export type AstToken = {
  path: string;
  value: ValueWithMode;
};

export type AstData = {
  metadata: AstMetadata;
  tokens: Map<string, ValueWithMode>;
};

export type Ast = {
  dataByType: { [tokenType in SpecifyDesignTokenTypeName]?: AstData };
  collectionModes: { [collectionName: string]: Array<string> };
  collectionPaths: { [collectionName: string]: Array<string> };
};

export type PrimitiveValue =
  | string
  | number
  | Array<string | number | Record<string, string | number | Array<string | number>>>
  | Record<string, { value: string | number | Array<string | number>; isCode?: true }>;

export type Value = { value: PrimitiveValue; isCode?: true };

export type Matcher = jsonValueMatcher<Value | undefined>;

export function tokenToAstTokenValue(token: TokenState, matcher: Matcher): AstToken | undefined {
  if (!token.isFullyResolvable) {
    return undefined;
  }

  const outputByMode = token.modes.reduce((acc, mode) => {
    const output = matchJsonValue<Value | undefined>(
      matcher,
      _ => {
        return undefined;
      },
      token,
      mode,
    );

    if (!output) return acc;

    acc[mode] = output;

    return acc;
  }, {} as { [mode: string]: Value });

  if (Object.values(outputByMode).every(v => !v)) return undefined;

  return { path: token.path.toString(), value: outputByMode };
}
