// @specifyapp/specify-design-token-format/src/definitions/designTokenTypeNames.ts
export const specifyJSONStringTypeName = 'string';
export const specifyJSONNumberTypeName = 'number';
export const specifyJSONBooleanTypeName = 'boolean';
export const specifyJSONNullTypeName = 'null';
export const specifyJSONArrayTypeName = 'array';
export const specifyJSONObjectTypeName = 'object';

export const specifyIntegerNumberTypeName = 'integerNumber';
export const specifyZeroToOneNumberTypeName = 'zeroToOneNumber';
export const specifyArcDegreeNumberTypeName = 'arcDegreeNumber';
export const specifyRGBColorNumberTypeName = 'rgbColorNumber';
export const specifyPositiveNumberTypeName = 'positiveNumber';
export const specifyPositiveIntegerNumberTypeName = 'positiveIntegerNumber';
export const specifyPercentageNumberTypeName = 'percentageNumber';

export const specifyHexadecimalColorStringTypeName = 'hexadecimalColorString';

export const specifyBitmapTypeName = 'bitmap';

export const specifyBitmapsTypeName = 'bitmaps';

export const specifyBitmapFormatTypeName = 'bitmapFormat';

export const specifyBlurTypeName = 'blur';

export const specifyBorderTypeName = 'border';

export const specifyBorderStyleTypeName = 'borderStyle';

export const specifyBorderStyleLineCapTypeName = 'borderStyleLineCap';

export const specifyBreakpointTypeName = 'breakpoint';

export const specifyColorTypeName = 'color';

export const specifyCubicBezierTypeName = 'cubicBezier';

export const specifyDimensionTypeName = 'dimension';

export const specifyDimensionUnitTypeName = 'dimensionUnit';

export const specifyDurationTypeName = 'duration';

export const specifyDurationUnitTypeName = 'durationUnit';

export const specifyFontTypeName = 'font';

export const specifyFontFamilyTypeName = 'fontFamily';

export const specifyFontFeatureTypeName = 'fontFeature';

export const specifyFontFeaturesTypeName = 'fontFeatures';

export const specifyFontFormatTypeName = 'fontFormat';

export const specifyFontStyleTypeName = 'fontStyle';

export const specifyFontWeightTypeName = 'fontWeight';

export const specifyGradientTypeName = 'gradient';

export const specifyGradientsTypeName = 'gradients';

export const specifyOpacityTypeName = 'opacity';

export const specifyRadiusTypeName = 'radius';

export const specifyRadiiTypeName = 'radii';

export const specifyShadowTypeName = 'shadow';

export const specifyShadowsTypeName = 'shadows';

export const specifyShadowTypeTypeName = 'shadowType';

export const specifySpacingTypeName = 'spacing';

export const specifySpacingsTypeName = 'spacings';

export const specifyStepsTimingFunctionTypeName = 'stepsTimingFunction';

export const specifyTextAlignHorizontalTypeName = 'textAlignHorizontal';

export const specifyTextAlignVerticalTypeName = 'textAlignVertical';

export const specifyTextDecorationTypeName = 'textDecoration';

export const specifyTextStyleTypeName = 'textStyle';

export const specifyTextTransformTypeName = 'textTransform';

export const specifyTransitionTypeName = 'transition';

export const specifyVectorTypeName = 'vector';

export const specifyVectorsTypeName = 'vectors';

export const specifyVectorFormatTypeName = 'vectorFormat';

export const specifyZIndexTypeName = 'zIndex';
