// @specifyapp/sdk/src/builtInParsers/parsers-generation/post-replace-string/definition.ts
import { z } from 'zod';
import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';

export const postReplaceStringParserName = 'post-replace-string';

export const postReplaceStringParserOptionsSchema = z
  .object({
    matchFileContent: z.union([
      z.string(),
      z.object({
        pattern: z.string(),
        flags: z.string().optional(),
      }),
    ]),
    replaceBy: z.string(),
    matchOutputFilePath: z.string().optional(),
  })
  .strict();

export type PostReplaceStringParserOptions = z.infer<typeof postReplaceStringParserOptionsSchema>;

export const postReplaceStringParserDefinition = createBuiltInParserDefinition({
  name: postReplaceStringParserName,
  kind: 'generation',
  hasOptionalOptions: false,
  optionsSchema: postReplaceStringParserOptionsSchema,
  hasOptionalOutput: true,
  outputTypes: undefined,
  inTypes: ['SDTF', 'SDTF Engine', 'vector', 'JSON', 'bitmap'],
  outType: 'SDTF Engine',
});

export type PostReplaceStringParserParserDefinition = typeof postReplaceStringParserDefinition;
