// @specifyapp/sdk/dist/builtInParsers/shared/to-scss-mixin-text-style/template.js
import { SpecifyError, specifyErrors } from '../../../errors/index.js';
function sanitizeName(name) {
    return name.replace(/[\/\| \\~,@#%^&*(){}|[\]?<>"'+=!]/g, '-');
}
var Variable;
(function (Variable) {
    Variable["MODE"] = "mode";
    Variable["PATH"] = "path";
    Variable["TOKEN"] = "token";
})(Variable || (Variable = {}));
export const DEFAULT_TEMPLATE = '{{path}}-{{token}}-{{mode}}';
export const DEFAULT_TEMPLATE_WITHOUT_MODE = '{{path}}-{{token}}';
const parse = (template) => (template.match(/{{(.*?)}}/g) ?? []).map(v => v.replace(/{|}/g, ''));
export const validateTemplate = (template) => parse(template).forEach(variable => {
    if (![Variable.PATH, Variable.MODE, Variable.TOKEN].includes(variable))
        throw new SpecifyError({
            errorKey: specifyErrors.PARSERS_ENGINE_INVALID_OPTION.errorKey,
            publicMessage: `'${variable}' is not a valid variable for the 'template' option.`,
        });
});
export const dataOfToken = (token, mode) => {
    const path = token.path.toArray().slice(0, token.path.length - 1);
    return {
        path,
        mode,
        token: token.name,
    };
};
export const renderTemplate = (template, data) => {
    const variables = parse(template);
    let rendered = template === DEFAULT_TEMPLATE && !data.mode ? DEFAULT_TEMPLATE_WITHOUT_MODE : template;
    for (let i = 0; i < variables.length; ++i) {
        const variable = variables[i];
        const regexp = new RegExp(`{{${variable}}}`);
        const currentName = data[variable];
        const name = Array.isArray(currentName) ? currentName.join('-') : currentName ?? '';
        const sanitized = sanitizeName(name);
        rendered = rendered.replace(regexp, sanitized);
    }
    // Remove starting - if it exists
    return rendered.replace(/^-/, '');
};
export const makeRenderer = (template) => (data) => {
    return renderTemplate(template, data);
};
