// @specifyapp/sdk/dist/parsersEngine/definitions/parsersEngineDataBox.js
import { z } from 'zod';
import { createSDTFEngine, specifyVectorFormatSchema, specifyBitmapFormatValuesSchema, specifyFontFormatSchema, specifyVectorProviderSchema, specifyBitmapProviderSchema, specifyFontProviderSchema, } from '@specifyapp/specify-design-token-format';
import { SpecifyError, specifyErrors } from '../../errors/index.js';
/* v8 ignore start */
export const SDTFDataBoxSchema = z
    .object({
    type: z.literal('SDTF'),
    // TODO @Nico: rename to tokenTree
    graph: z.custom(value => value),
    metadata: z.custom(value => value).optional(),
})
    .strict();
export const SDTFEngineDataBoxSchema = z
    .object({
    type: z.literal('SDTF Engine'),
    engine: z.custom(),
})
    .strict();
export const specifyRepositoryDataBoxSchema = z
    .object({
    type: z.literal('repository'),
    owner: z.string().startsWith('@'),
    name: z.string(),
})
    .strict();
export const JSONDataBoxSchema = z
    .object({
    type: z.literal('JSON'),
    json: z.union([
        z.record(z.string(), z.unknown()),
        z.array(z.unknown()),
        z.string(),
        z.number(),
        z.boolean(),
        z.null(),
    ]),
})
    .strict();
export const CustomDataBoxSchema = z
    .object({
    type: z.literal('custom'),
    data: z.unknown(),
})
    .strict();
function makeAssetDataBoxSchema() {
    return ({ type, format, extension, provider, } = {}) => {
        return z
            .object({
            type: z.literal((type ?? 'asset')),
            assets: z.array(z
                .object({
                path: z.array(z.string()),
                mode: z.string(),
                url: z.string(),
                provider: (provider ?? z.string()),
                format: (format ?? z.string()),
                token: z.custom(value => value),
                ...(extension ?? {}),
            })
                .strict()),
        })
            .strict();
    };
}
export const VectorDataBoxSchema = makeAssetDataBoxSchema()({
    type: 'vector',
    format: specifyVectorFormatSchema,
    provider: specifyVectorProviderSchema,
    extension: {
        vector: z.string().optional(),
    },
});
export const BitmapDataBoxSchema = makeAssetDataBoxSchema()({
    type: 'bitmap',
    format: specifyBitmapFormatValuesSchema,
    provider: specifyBitmapProviderSchema,
    extension: {
        bitmap: z.array(z.number()).optional(),
    },
});
export const AssetDataBoxSchema = makeAssetDataBoxSchema()({
    type: undefined,
    format: z.union([
        specifyBitmapFormatValuesSchema,
        specifyVectorFormatSchema,
        specifyFontFormatSchema,
    ]),
    provider: z.union([
        specifyBitmapProviderSchema,
        specifyVectorProviderSchema,
        specifyFontProviderSchema,
    ]),
});
export const parsersEngineDataBoxSchema = z.discriminatedUnion('type', [
    SDTFDataBoxSchema,
    SDTFEngineDataBoxSchema,
    specifyRepositoryDataBoxSchema,
    JSONDataBoxSchema,
    CustomDataBoxSchema,
    VectorDataBoxSchema,
    AssetDataBoxSchema,
    BitmapDataBoxSchema,
]);
export const parsersEngineDataBoxSchemaMap = {
    SDTF: SDTFDataBoxSchema,
    'SDTF Engine': SDTFEngineDataBoxSchema,
    repository: specifyRepositoryDataBoxSchema,
    JSON: JSONDataBoxSchema,
    custom: CustomDataBoxSchema,
    vector: VectorDataBoxSchema,
    asset: AssetDataBoxSchema,
    bitmap: BitmapDataBoxSchema,
};
export function getParsersEngineDataBoxUnionSchema(types) {
    if (types.length === 0)
        throw new Error('At least one type must be provided');
    if (types.length === 1)
        return parsersEngineDataBoxSchemaMap[types[0]];
    return z.union(types.map(type => parsersEngineDataBoxSchemaMap[type]));
}
export const serializedParsersEngineDataBoxSchema = z.discriminatedUnion('type', [
    SDTFDataBoxSchema,
    specifyRepositoryDataBoxSchema,
    JSONDataBoxSchema,
    CustomDataBoxSchema,
    VectorDataBoxSchema,
    AssetDataBoxSchema,
    BitmapDataBoxSchema,
]);
/* v8 ignore stop */
export function getDefaultParsersEngineDataBox(type) {
    switch (type) {
        case 'SDTF Engine': {
            const t = {
                type: 'SDTF Engine',
                engine: createSDTFEngine(),
            };
            return t;
        }
        case 'SDTF':
        case undefined: {
            const t = {
                type: 'SDTF',
                graph: {},
            };
            return t;
        }
        case 'repository': {
            const t = {
                type: 'repository',
                owner: '',
                name: '',
            };
            return t;
        }
        case 'JSON': {
            const t = {
                type: 'JSON',
                json: {},
            };
            return t;
        }
        case 'bitmap': {
            const t = {
                type: 'bitmap',
                assets: [],
            };
            return t;
        }
        case 'vector': {
            const t = {
                type: 'vector',
                assets: [],
            };
            return t;
        }
        case 'asset': {
            const t = {
                type: 'asset',
                assets: [],
            };
            return t;
        }
        case 'custom': {
            const t = {
                type: 'custom',
                data: null,
            };
            return t;
        }
        default: {
            const _exhaustiveCheck = type;
            throw new SpecifyError({
                errorKey: specifyErrors.PARSERS_ENGINE_UNKNOWN_ERROR.errorKey,
                publicMessage: `Unknown data box type "${type}".`,
            });
        }
    }
}
