// @specifyapp/sdk/dist/parsersEngine/definitions/parserOutput.js
import { z } from 'zod';
export const textOutputSchema = z
    .object({
    type: z.literal('text'),
    text: z.string(),
})
    .strict();
export const sdtfOutputSchema = z
    .object({
    type: z.literal('SDTF'),
    graph: z.custom(value => value),
})
    .strict();
export const jsonOutputSchema = z
    .object({
    type: z.literal('JSON'),
    json: z.union([
        z.record(z.string(), z.unknown()),
        z.array(z.unknown()),
        z.string(),
        z.number(),
        z.boolean(),
        z.null(),
    ]),
})
    .strict();
export const filesOutputFileSchema = z
    .object({
    path: z.string(),
    content: z.union([
        z.object({
            type: z.literal('text'),
            text: z.string(),
        }),
        z.object({
            type: z.literal('url'),
            url: z.string(),
        }),
    ]),
})
    .strict();
export const filesOutputSchema = z
    .object({
    type: z.literal('files'),
    files: z.array(filesOutputFileSchema),
})
    .strict();
export const parserOutputSchema = z.discriminatedUnion('type', [
    textOutputSchema,
    sdtfOutputSchema,
    jsonOutputSchema,
    filesOutputSchema,
]);
