// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-tailwind/definition.ts
import { z } from 'zod';
import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';

export const toTailwindParserName = 'to-tailwind';

export const toTailwindParserDefinition = createBuiltInParserDefinition({
  name: toTailwindParserName,
  kind: 'generation',
  hasOptionalOptions: true,
  optionsSchema: z
    .object({
      useCssVariable: z.boolean().optional(),
      cssVariableTemplate: z
        .object({
          tokenNameTemplate: z.string().optional(),
          tokenNotInCollectionNameTemplate: z.string().optional(),
        })
        .strict()
        .optional(),
      removeModesIfSameValue: z.boolean().optional(),
      removeSingleMode: z.boolean().optional(),
    })
    .strict(),
  hasOptionalOutput: false,
  outputTypes: ['file'],
  inTypes: ['SDTF', 'SDTF Engine'],
  outType: 'SDTF Engine',
});

export type ToTailwindParserDefinition = typeof toTailwindParserDefinition;
export type ToTailwindParserOptions = z.infer<ToTailwindParserDefinition['optionsSchema']>;
