// @specifyapp/sdk/dist/httpApi/internals/httpResponseBody.js
import { specifyErrors } from '../../errors/index.js';
export function makeHttpSuccessResponseBody(messages, payload) {
    return {
        status: 'success',
        messages,
        payload,
    };
}
export function makeHttpErrorResponseBody(messages) {
    return {
        status: 'error',
        messages,
    };
}
export function matchIsHttpErrorResponseBody(candidate) {
    return (typeof candidate === 'object' &&
        candidate !== null &&
        candidate.status === 'error' &&
        'messages' in candidate &&
        Array.isArray(candidate.messages));
}
export function getLastHttpResponseErrorMessage(responseBody) {
    if (Array.isArray(responseBody.messages) === false)
        throw new Error('Expected responseBody.messages to be an array');
    const maybeError = responseBody.messages[responseBody.messages.length - 1];
    if (maybeError?.type === 'error') {
        return maybeError;
    }
    return {
        type: 'error',
        content: maybeError?.content ? `"${maybeError?.content}" is not an error` : 'Unknown error',
        errorKey: specifyErrors.UNEXPECTED_ERROR.errorKey,
    };
}
export function getHttpResponseWarningAndErrorMessages(responseBody) {
    if (Array.isArray(responseBody.messages) === false)
        throw new Error('Expected responseBody.messages to be an array');
    return responseBody.messages.filter(m => m.type === 'warning' || m.type === 'error');
}
