// @specifyapp/sdk/dist/builtInParsers/parsers-generation/generate-barrel-file/definition.js
import { z } from 'zod';
import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';
export const generateBarrelFileParserName = 'generate-barrel-file';
export const generateBarrelFileParserOptionsSchema = z
    .object({
    useSemicolon: z.boolean().optional(),
    useFileExtensionsInImport: z.boolean().optional(),
    quoteCharacter: z.union([z.literal("'"), z.literal('"')]).optional(),
    removePatternInOutputFilePath: z.string().optional(),
    matchOutputFilePathPattern: z.string().optional(),
    fileExtensionsMapping: z.record(z.string(), z.string()).optional(),
})
    .strict();
export const generateBarrelFileParserDefinition = createBuiltInParserDefinition({
    name: generateBarrelFileParserName,
    kind: 'generation',
    hasOptionalOptions: true,
    optionsSchema: generateBarrelFileParserOptionsSchema,
    hasOptionalOutput: false,
    outputTypes: ['file'],
    inTypes: ['SDTF', 'SDTF Engine', 'vector', 'bitmap', 'asset'],
    outType: 'JSON',
});
