// @specifyapp/specify-design-token-format/dist/engine/createSDTFEngine.js
import { TreeState } from './state/TreeState.js';
import { makeRunQuery } from './query/index.js';
import { addCollectionMutationDefinition, addGroupMutationDefinition, addTokenMutationDefinition, createTokenModeValueMutationDefinition, deleteAllViewsMutationDefinition, deleteCollectionModeMutationDefinition, deleteCollectionMutationDefinition, deleteGroupMutationDefinition, deleteTokenModeValueMutationDefinition, deleteTokenMutationDefinition, deleteViewMutationDefinition, loadTokenTreeMutationDefinition, moveCollectionMutationDefinition, moveGroupMutationDefinition, moveTokenMutationDefinition, registerViewMutationDefinition, renameCollectionModeMutationDefinition, renameCollectionMutationDefinition, renameGroupMutationDefinition, renameNodeMutationDefinition, renameTokenModeMutationDefinition, renameTokenMutationDefinition, resetTokenTreeMutationDefinition, resolveTokenValueAliasesMutationDefinition, setActiveViewMutationDefinition, truncateCollectionMutationDefinition, truncateGroupMutationDefinition, updateCollectionDescriptionMutationDefinition, updateCollectionExtensionsMutationDefinition, updateGroupDescriptionMutationDefinition, updateGroupExtensionsMutationDefinition, updateTokenDescriptionMutationDefinition, updateTokenExtensionsMutationDefinition, updateTokenModeValueMutationDefinition, updateTokenValueMutationDefinition, updateViewMutationDefinition, } from './mutations/definitions.js';
import { sdtfEngineSerializedMetadataSchema, } from './SDTFEngineSerializedState.js';
import { AliasReferenceSet, } from './state/AliasReferenceSet.js';
import { fillTreeNodesStateAndAliasReferences } from './builder/createTreeState.js';
import { analyzeTokenTree } from './parser/analyzeTokenTree.js';
import { ViewsState } from './state/ViewsState.js';
import { TreeNodesState } from './state/TreeNodesState.js';
function makeEngine(analyzedTokenTree, parsedMetadata) {
    const globalState = new TreeNodesState();
    const aliasReferences = new AliasReferenceSet();
    const viewsState = new ViewsState();
    const treeState = new TreeState({
        globalState,
        aliasReferences,
        viewsState,
        activeViewName: null, // set to null before registering views
    });
    fillTreeNodesStateAndAliasReferences(globalState, aliasReferences, analyzedTokenTree, treeState);
    // Register views
    parsedMetadata.views.forEach(view => {
        viewsState.register(view.name, view.query, treeState);
    });
    // Register active view once all views are registered
    if (parsedMetadata.activeViewName !== null) {
        treeState.setActiveView(parsedMetadata?.activeViewName);
    }
    const runQuery = makeRunQuery(treeState);
    function loadTokenTree(tokenTree) {
        const newAnalyzedTokenTree = analyzeTokenTree(tokenTree);
        globalState.clear();
        aliasReferences.clear();
        fillTreeNodesStateAndAliasReferences(globalState, aliasReferences, newAnalyzedTokenTree, treeState);
        viewsState.updateAll(treeState);
    }
    function clone() {
        return makeEngine(globalState.toAnalyzedTokenTree(), treeState.exportMetadata());
    }
    /* v8 ignore start */
    return {
        renderJSONTree: (...args) => treeState.renderJSONTree(...args),
        mutation: {
            resetTokenTree: resetTokenTreeMutationDefinition.remapAndPipeWith(
            // @ts-expect-error - resetTokenTreeMutationDefinition has no arguments
            () => [], () => treeState.reset()),
            loadTokenTree: loadTokenTreeMutationDefinition.remapAndPipeWith(({ tokens }) => [tokens], loadTokenTree),
            // Tree views
            registerView: registerViewMutationDefinition.remapAndPipeWith(({ name, query, shouldSetActive }) => [
                name,
                query,
                shouldSetActive === null ? false : shouldSetActive,
            ], 
            // @ts-expect-error
            (...args) => treeState.registerView(...args)),
            updateView: updateViewMutationDefinition.remapAndPipeWith(({ name, query, shouldSetActive }) => [
                name,
                query,
                shouldSetActive === null ? false : shouldSetActive,
            ], 
            // @ts-expect-error
            (...args) => treeState.updateView(...args)),
            setActiveView: setActiveViewMutationDefinition.remapAndPipeWith(({ name }) => [name], 
            // @ts-expect-error
            (...args) => treeState.setActiveView(...args)),
            deleteView: deleteViewMutationDefinition.remapAndPipeWith(({ name }) => [name], (...args) => {
                // @ts-expect-error
                return treeState.deleteView(...args);
            }),
            deleteAllViews: deleteAllViewsMutationDefinition.remapAndPipeWith(
            // @ts-expect-error - deleteAllViewsMutationDefinition has no arguments
            () => [], () => treeState.deleteAllViews()),
            // Collection
            addCollection: addCollectionMutationDefinition.remapAndPipeWith(({ parentPath, name, collectionProperties }) => [
                parentPath.append(name),
                collectionProperties,
            ], 
            // @ts-expect-error
            (...args) => treeState.addCollection(...args)),
            renameCollection: renameCollectionMutationDefinition.remapAndPipeWith(({ atPath, name }) => [atPath, name], 
            // @ts-expect-error
            (...args) => treeState.renameCollection(...args)),
            updateCollectionDescription: updateCollectionDescriptionMutationDefinition.remapAndPipeWith(({ atPath, description }) => [atPath, description], 
            // @ts-expect-error
            (...args) => treeState.updateCollectionDescription(...args)),
            updateCollectionExtensions: updateCollectionExtensionsMutationDefinition.remapAndPipeWith(({ atPath, extensions }) => [atPath, extensions], 
            // @ts-expect-error
            (...args) => treeState.updateCollectionExtensions(...args)),
            renameCollectionMode: renameCollectionModeMutationDefinition.remapAndPipeWith(({ atPath, fromMode, toMode }) => [atPath, fromMode, toMode], 
            // @ts-expect-error
            (...args) => treeState.renameCollectionMode(...args)),
            deleteCollectionMode: deleteCollectionModeMutationDefinition.remapAndPipeWith(({ atPath, mode }) => [atPath, mode], 
            // @ts-expect-error
            (...args) => treeState.deleteCollectionMode(...args)),
            truncateCollection: truncateCollectionMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.truncateCollection(...args)),
            deleteCollection: deleteCollectionMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.deleteCollection(...args)),
            moveCollection: moveCollectionMutationDefinition.remapAndPipeWith(({ atPath, toPath }) => [atPath, toPath], 
            // @ts-expect-error
            (...args) => treeState.moveCollection(...args)),
            // Group
            addGroup: addGroupMutationDefinition.remapAndPipeWith(({ parentPath, name, groupProperties }) => [parentPath.append(name), groupProperties], 
            // @ts-expect-error
            (...args) => treeState.addGroup(...args)),
            renameGroup: renameGroupMutationDefinition.remapAndPipeWith(({ atPath, name }) => [atPath, name], 
            // @ts-expect-error
            (...args) => treeState.renameGroup(...args)),
            updateGroupDescription: updateGroupDescriptionMutationDefinition.remapAndPipeWith(({ atPath, description }) => [atPath, description], 
            // @ts-expect-error
            (...args) => treeState.updateGroupDescription(...args)),
            updateGroupExtensions: updateGroupExtensionsMutationDefinition.remapAndPipeWith(({ atPath, extensions }) => [atPath, extensions], 
            // @ts-expect-error
            (...args) => treeState.updateGroupExtensions(...args)),
            truncateGroup: truncateGroupMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.truncateGroup(...args)),
            deleteGroup: deleteGroupMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.deleteGroup(...args)),
            moveGroup: moveGroupMutationDefinition.remapAndPipeWith(({ atPath, toPath }) => [atPath, toPath], 
            // @ts-expect-error
            (...args) => treeState.moveGroup(...args)),
            // Token
            addToken: addTokenMutationDefinition.remapAndPipeWith(({ parentPath, name, tokenProperties }) => [parentPath.append(name), tokenProperties], 
            // @ts-expect-error
            (...args) => treeState.addToken(...args)),
            renameToken: renameTokenMutationDefinition.remapAndPipeWith(({ atPath, name }) => [atPath, name], 
            // @ts-expect-error
            (...args) => treeState.renameToken(...args)),
            updateTokenDescription: updateTokenDescriptionMutationDefinition.remapAndPipeWith(({ atPath, description }) => [atPath, description], 
            // @ts-expect-error
            (...args) => treeState.updateTokenDescription(...args)),
            updateTokenExtensions: updateTokenExtensionsMutationDefinition.remapAndPipeWith(({ atPath, extensions }) => [atPath, extensions], 
            // @ts-expect-error
            (...args) => treeState.updateTokenExtensions(...args)),
            updateTokenValue: updateTokenValueMutationDefinition.remapAndPipeWith(({ atPath, value }) => [atPath, value], 
            // @ts-expect-error
            (...args) => treeState.updateTokenValue(...args)),
            resolveTokenValueAliases: resolveTokenValueAliasesMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.resolveTokenValueAliases(...args)),
            updateTokenModeValue: updateTokenModeValueMutationDefinition.remapAndPipeWith(({ atPath, mode, value }) => [atPath, mode, value], 
            // @ts-expect-error
            (...args) => treeState.updateTokenModeValue(...args)),
            renameTokenMode: renameTokenModeMutationDefinition.remapAndPipeWith(({ atPath, fromMode, toMode }) => [atPath, fromMode, toMode], 
            // @ts-expect-error
            (...args) => treeState.renameTokenMode(...args)),
            createTokenModeValue: createTokenModeValueMutationDefinition.remapAndPipeWith(({ atPath, mode, value }) => [atPath, mode, value], 
            // @ts-expect-error
            (...args) => treeState.createTokenModeValue(...args)),
            deleteTokenModeValue: deleteTokenModeValueMutationDefinition.remapAndPipeWith(({ atPath, mode }) => [atPath, mode], 
            // @ts-expect-error
            (...args) => treeState.deleteTokenModeValue(...args)),
            deleteToken: deleteTokenMutationDefinition.remapAndPipeWith(({ atPath }) => [atPath], 
            // @ts-expect-error
            (...args) => treeState.deleteToken(...args)),
            moveToken: moveTokenMutationDefinition.remapAndPipeWith(({ atPath, toPath }) => [atPath, toPath], 
            // @ts-expect-error
            (...args) => treeState.moveToken(...args)),
            renameNode: renameNodeMutationDefinition.remapAndPipeWith(({ atPath, name }) => [atPath, name], 
            // @ts-expect-error
            (...args) => treeState.renameNode(...args)),
        },
        query: {
            run: runQuery,
            // @ts-expect-error
            getTokenState: (...args) => treeState.getTokenState(...args),
            // @ts-expect-error
            getGroupState: (...args) => treeState.getGroupState(...args),
            // @ts-expect-error
            getCollectionState: (...args) => treeState.getCollectionState(...args),
            getNearestCollectionState: (...args) => 
            // @ts-expect-error
            treeState.getNearestCollectionState(...args),
            getAllTokenStates: (...args) => treeState.getAllTokenStates(...args),
            getAllGroupStates: (...args) => treeState.getAllGroupStates(...args),
            getAllCollectionStates: (...args) => treeState.getAllCollectionStates(...args),
            getAllNodeStates: (...args) => treeState.getAllNodeStates(...args),
            // @ts-expect-error
            getTokenChildrenOf: (...args) => treeState.getTokenChildrenOf(...args),
            // @ts-expect-error
            getGroupChildrenOf: (...args) => treeState.getGroupChildrenOf(...args),
            // @ts-expect-error
            getCollectionChildrenOf: (...args) => treeState.getCollectionChildrenOf(...args),
            // @ts-expect-error
            getChildrenOf: (...args) => treeState.getChildrenOf(...args),
            // @ts-expect-error
            getParentsOf: (...args) => treeState.getParentsOf(...args),
            getGroupChildren: (...args) => treeState.getGroupChildren(...args),
            getTokenChildren: (...args) => treeState.getTokenChildren(...args),
            getCollectionChildren: (...args) => treeState.getCollectionChildren(...args),
            renderJSONTree: (...args) => treeState.renderJSONTree(...args),
            // @ts-expect-error
            getAliasReference: (...args) => treeState.getAliasReference(...args),
            // @ts-expect-error
            getAllAliasReferences: (...args) => treeState.getAllAliasReferences(...args),
            // @ts-expect-error
            getAliasReferencesTo: (...args) => treeState.getAliasReferencesTo(...args),
            // @ts-expect-error
            getAliasReferencesFrom: (...args) => treeState.getAliasReferencesFrom(...args),
            getStatefulAliasReference: (...args) => 
            // @ts-expect-error
            treeState.getStatefulAliasReference(...args),
            getStatefulAliasReferencesTo: (...args) => 
            // @ts-expect-error
            treeState.getStatefulAliasReferencesTo(...args),
            getStatefulAliasReferencesFrom: (...args) => 
            // @ts-expect-error
            treeState.getStatefulAliasReferencesFrom(...args),
            // @ts-expect-error
            listViews: (...args) => treeState.listViews(...args),
            // @ts-expect-error
            getActiveView: (...args) => treeState.getActiveView(...args),
        },
        exportEngineState: () => treeState.exportAll(),
        clone,
    };
    /* v8 ignore stop */
}
export function createSDTFEngine(tokenTree, metadata) {
    const analyzedTokenTree = analyzeTokenTree(tokenTree);
    const parsedMetadata = sdtfEngineSerializedMetadataSchema.optional().parse(metadata);
    return makeEngine(analyzedTokenTree, parsedMetadata ?? {
        activeViewName: null,
        views: [],
    });
}
export function isSDTFEngine(data) {
    if (typeof data !== 'object' || data === null)
        return false;
    const hasRenderFn = 'renderJSONTree' in data && typeof data.renderJSONTree === 'function';
    const hasMutationResetFn = 'mutation' in data &&
        'resetTokenTree' in data.mutation &&
        typeof data.mutation.resetTokenTree === 'function';
    const hasQueryRunFn = 'query' in data && 'run' in data.query && typeof data.query.run === 'function';
    return hasRenderFn && hasMutationResetFn && hasQueryRunFn;
}
