// @specifyapp/sdk/dist/builtInParsers/parsers-utility/convert-dimension/conversion.js
import { PX_TO_CM_RATIO, PX_TO_IN_RATIO, PX_TO_MM_RATIO, PX_TO_PC_RATIO, PX_TO_PT_RATIO, REM_DEFAULT_BASE, } from './constants.js';
// Rounding to 3 decimals
export function roundNumber(v) {
    return Math.round((v + Number.EPSILON) * 1000) / 1000;
}
export function pxToCm(px) {
    return px * PX_TO_CM_RATIO;
}
export function pxToMm(px) {
    return px * PX_TO_MM_RATIO;
}
export function pxToIn(px) {
    return px * PX_TO_IN_RATIO;
}
export function pxToPt(px) {
    return px * PX_TO_PT_RATIO;
}
export function pxToPc(px) {
    return px * PX_TO_PC_RATIO;
}
export function pxToRem(px, baseValue = REM_DEFAULT_BASE) {
    return px / baseValue;
}
export function cmToPx(cm) {
    return cm / PX_TO_CM_RATIO;
}
export function mmToPx(mm) {
    return mm / PX_TO_MM_RATIO;
}
export function inToPx(in_) {
    return in_ / PX_TO_IN_RATIO;
}
export function ptToPx(pt) {
    return pt / PX_TO_PT_RATIO;
}
export function pcToPx(pc) {
    return pc / PX_TO_PC_RATIO;
}
export function remToPx(rem, baseValue = REM_DEFAULT_BASE) {
    return rem * baseValue;
}
// TIPS: If you want to support a new unit, I recommend to take the type below, set it to a new object with the Record type
// so your IDE can generate all the cases for you, and then you pick what you're interested in.
// Example: const test: Record<`${SpecifyDimensionUnitValue}:${SpecifyDimensionUnitValue}`, (value: number, baseValue: baseValue) => number> = {}
export const converter = {
    'null:%': function (value, _baseValue) {
        return value;
    },
    'null:px': function (value, _baseValue) {
        return value;
    },
    'null:em': function (value, _baseValue) {
        return value;
    },
    'null:rem': function (value, _baseValue) {
        return value;
    },
    'null:pt': function (value, _baseValue) {
        return value;
    },
    'null:pc': function (value, _baseValue) {
        return value;
    },
    'null:in': function (value, _baseValue) {
        return value;
    },
    'null:cm': function (value, _baseValue) {
        return value;
    },
    'null:mm': function (value, _baseValue) {
        return value;
    },
    'null:ex': function (value, _baseValue) {
        return value;
    },
    'null:cap': function (value, _baseValue) {
        return value;
    },
    'null:ch': function (value, _baseValue) {
        return value;
    },
    'null:ic': function (value, _baseValue) {
        return value;
    },
    'null:lh': function (value, _baseValue) {
        return value;
    },
    'null:rlh': function (value, _baseValue) {
        return value;
    },
    'null:vw': function (value, _baseValue) {
        return value;
    },
    'null:svw': function (value, _baseValue) {
        return value;
    },
    'null:lvw': function (value, _baseValue) {
        return value;
    },
    'null:dvw': function (value, _baseValue) {
        return value;
    },
    'null:vh': function (value, _baseValue) {
        return value;
    },
    'null:svh': function (value, _baseValue) {
        return value;
    },
    'null:lvh': function (value, _baseValue) {
        return value;
    },
    'null:dvh': function (value, _baseValue) {
        return value;
    },
    'null:vi': function (value, _baseValue) {
        return value;
    },
    'null:svi': function (value, _baseValue) {
        return value;
    },
    'null:lvi': function (value, _baseValue) {
        return value;
    },
    'null:dvi': function (value, _baseValue) {
        return value;
    },
    'null:vb': function (value, _baseValue) {
        return value;
    },
    'null:svb': function (value, _baseValue) {
        return value;
    },
    'null:lvb': function (value, _baseValue) {
        return value;
    },
    'null:dvb': function (value, _baseValue) {
        return value;
    },
    'null:vmin': function (value, _baseValue) {
        return value;
    },
    'null:svmin': function (value, _baseValue) {
        return value;
    },
    'null:lvmin': function (value, _baseValue) {
        return value;
    },
    'null:dvmin': function (value, _baseValue) {
        return value;
    },
    'null:vmax': function (value, _baseValue) {
        return value;
    },
    'null:svmax': function (value, _baseValue) {
        return value;
    },
    'null:lvmax': function (value, _baseValue) {
        return value;
    },
    'null:dvmax': function (value, _baseValue) {
        return value;
    },
    '%:vw': function (value, _baseValue) {
        return value;
    },
    '%:svw': function (value, _baseValue) {
        return value;
    },
    '%:lvw': function (value, _baseValue) {
        return value;
    },
    '%:dvw': function (value, _baseValue) {
        return value;
    },
    '%:vh': function (value, _baseValue) {
        return value;
    },
    '%:svh': function (value, _baseValue) {
        return value;
    },
    '%:lvh': function (value, _baseValue) {
        return value;
    },
    '%:dvh': function (value, _baseValue) {
        return value;
    },
    '%:vi': function (value, _baseValue) {
        return value;
    },
    '%:svi': function (value, _baseValue) {
        return value;
    },
    '%:lvi': function (value, _baseValue) {
        return value;
    },
    '%:dvi': function (value, _baseValue) {
        return value;
    },
    '%:vb': function (value, _baseValue) {
        return value;
    },
    '%:svb': function (value, _baseValue) {
        return value;
    },
    '%:lvb': function (value, _baseValue) {
        return value;
    },
    '%:dvb': function (value, _baseValue) {
        return value;
    },
    '%:vmin': function (value, _baseValue) {
        return value;
    },
    '%:svmin': function (value, _baseValue) {
        return value;
    },
    '%:lvmin': function (value, _baseValue) {
        return value;
    },
    '%:dvmin': function (value, _baseValue) {
        return value;
    },
    '%:vmax': function (value, _baseValue) {
        return value;
    },
    '%:svmax': function (value, _baseValue) {
        return value;
    },
    '%:lvmax': function (value, _baseValue) {
        return value;
    },
    '%:dvmax': function (value, _baseValue) {
        return value;
    },
    'px:rem': function (value, baseValue) {
        return pxToRem(value, baseValue.rem);
    },
    'px:pt': function (value, _baseValue) {
        return pxToPt(value);
    },
    'px:pc': function (value, _baseValue) {
        return pxToPc(value);
    },
    'px:in': function (value, _baseValue) {
        return pxToIn(value);
    },
    'px:cm': function (value, _baseValue) {
        return pxToCm(value);
    },
    'px:mm': function (value, _baseValue) {
        return pxToMm(value);
    },
    'rem:px': function (value, baseValue) {
        return remToPx(value, baseValue.rem);
    },
    'rem:pt': function (value, baseValue) {
        return pxToPt(remToPx(value, baseValue.rem));
    },
    'rem:pc': function (value, baseValue) {
        return pxToPc(remToPx(value, baseValue.rem));
    },
    'rem:in': function (value, baseValue) {
        return pxToIn(remToPx(value, baseValue.rem));
    },
    'rem:cm': function (value, baseValue) {
        return pxToCm(remToPx(value, baseValue.rem));
    },
    'rem:mm': function (value, baseValue) {
        return pxToMm(remToPx(value, baseValue.rem));
    },
    'pt:px': function (value, _baseValue) {
        return ptToPx(value);
    },
    'pt:rem': function (value, baseValue) {
        return pxToRem(ptToPx(value), baseValue.rem);
    },
    'pt:pc': function (value, _baseValue) {
        return pxToPc(ptToPx(value));
    },
    'pt:in': function (value, _baseValue) {
        return pxToIn(ptToPx(value));
    },
    'pt:cm': function (value, _baseValue) {
        return pxToCm(ptToPx(value));
    },
    'pt:mm': function (value, _baseValue) {
        return pxToMm(ptToPx(value));
    },
    'pc:px': function (value, _baseValue) {
        return pcToPx(value);
    },
    'pc:rem': function (value, baseValue) {
        return pxToRem(pcToPx(value), baseValue.rem);
    },
    'pc:pt': function (value, _baseValue) {
        return pxToPt(ptToPx(value));
    },
    'pc:in': function (value, _baseValue) {
        return pxToIn(ptToPx(value));
    },
    'pc:cm': function (value, _baseValue) {
        return pxToCm(ptToPx(value));
    },
    'pc:mm': function (value, _baseValue) {
        return pxToMm(ptToPx(value));
    },
    'in:px': function (value, _baseValue) {
        return inToPx(value);
    },
    'in:rem': function (value, baseValue) {
        return pxToRem(inToPx(value), baseValue.rem);
    },
    'in:pt': function (value, _baseValue) {
        return pxToPt(inToPx(value));
    },
    'in:pc': function (value, _baseValue) {
        return pxToPc(inToPx(value));
    },
    'in:cm': function (value, _baseValue) {
        return pxToCm(inToPx(value));
    },
    'in:mm': function (value, _baseValue) {
        return pxToMm(inToPx(value));
    },
    'cm:px': function (value, _baseValue) {
        return cmToPx(value);
    },
    'cm:rem': function (value, baseValue) {
        return pxToRem(cmToPx(value), baseValue.rem);
    },
    'cm:pt': function (value, _baseValue) {
        return pxToPt(cmToPx(value));
    },
    'cm:pc': function (value, _baseValue) {
        return pxToPc(cmToPx(value));
    },
    'cm:in': function (value, _baseValue) {
        return pxToIn(cmToPx(value));
    },
    'cm:mm': function (value, _baseValue) {
        return pxToMm(cmToPx(value));
    },
    'mm:px': function (value, _baseValue) {
        return mmToPx(value);
    },
    'mm:rem': function (value, baseValue) {
        return pxToRem(mmToPx(value), baseValue.rem);
    },
    'mm:pt': function (value, _baseValue) {
        return pxToPt(mmToPx(value));
    },
    'mm:pc': function (value, _baseValue) {
        return pxToPc(mmToPx(value));
    },
    'mm:in': function (value, _baseValue) {
        return pxToIn(mmToPx(value));
    },
    'mm:cm': function (value, _baseValue) {
        return pxToCm(mmToPx(value));
    },
    'vw:%': function (value, _baseValue) {
        return value;
    },
    'svw:%': function (value, _baseValue) {
        return value;
    },
    'lvw:%': function (value, _baseValue) {
        return value;
    },
    'dvw:%': function (value, _baseValue) {
        return value;
    },
    'vh:%': function (value, _baseValue) {
        return value;
    },
    'svh:%': function (value, _baseValue) {
        return value;
    },
    'lvh:%': function (value, _baseValue) {
        return value;
    },
    'dvh:%': function (value, _baseValue) {
        return value;
    },
    'vi:%': function (value, _baseValue) {
        return value;
    },
    'svi:%': function (value, _baseValue) {
        return value;
    },
    'lvi:%': function (value, _baseValue) {
        return value;
    },
    'dvi:%': function (value, _baseValue) {
        return value;
    },
    'vb:%': function (value, _baseValue) {
        return value;
    },
    'svb:%': function (value, _baseValue) {
        return value;
    },
    'lvb:%': function (value, _baseValue) {
        return value;
    },
    'dvb:%': function (value, _baseValue) {
        return value;
    },
    'vmin:%': function (value, _baseValue) {
        return value;
    },
    'svmin:%': function (value, _baseValue) {
        return value;
    },
    'lvmin:%': function (value, _baseValue) {
        return value;
    },
    'dvmin:%': function (value, _baseValue) {
        return value;
    },
    'vmax:%': function (value, _baseValue) {
        return value;
    },
    'svmax:%': function (value, _baseValue) {
        return value;
    },
    'lvmax:%': function (value, _baseValue) {
        return value;
    },
    'dvmax:%': function (value, _baseValue) {
        return value;
    },
};
export function convertDimension({ value, unit }, targetUnit, baseValue) {
    if (targetUnit === null) {
        return { unit: null, value };
    }
    else if (unit === null || unit === targetUnit) {
        return { unit: targetUnit, value };
    }
    const output = converter[`${unit}:${targetUnit}`]?.(value, baseValue);
    if (!output)
        return undefined;
    return { value: roundNumber(output), unit: targetUnit };
}
