// @specifyapp/sdk/dist/builtInParsers/parsers-utility/convert-dimension/updateToken.js
import { matchRawValue } from '../../utils/rawValueMatcher.js';
import { getDimensionUpdateValue } from './tokenTypes/dimension.js';
import { getBorderUpdateValue } from './tokenTypes/border.js';
import { getRadiiUpdateValue } from './tokenTypes/radii.js';
import { getShadowUpdateValue } from './tokenTypes/shadow.js';
import { getShadowsUpdateValue } from './tokenTypes/shadows.js';
import { getTextStyleUpdateValue } from './tokenTypes/textStyle.js';
export function updateToken(token, options, toolbox) {
    token
        .getStatefulValueResult()
        .resolveDeepValue()
        .mapTopLevelValue(value => value.resolveDeepValue().mapRawValueWithTokenState((value, mode, tokenState) => {
        const output = matchRawValue({
            border: border => getBorderUpdateValue(border, options, tokenState, mode, toolbox, options.applyToKeys.border),
            shadow: v => getShadowUpdateValue(v, options, tokenState, mode, toolbox, options.applyToKeys.shadow),
            shadows: v => getShadowsUpdateValue(v, options, tokenState, mode, toolbox, options.applyToKeys.shadow),
            textStyle: v => getTextStyleUpdateValue(v, options, tokenState, mode, toolbox, options.applyToKeys.textStyle),
            dimension: v => getDimensionUpdateValue(v, options, toolbox),
            spacing: v => getDimensionUpdateValue(v, options, toolbox),
            blur: v => getDimensionUpdateValue(v, options, toolbox),
            breakpoint: v => getDimensionUpdateValue(v, options, toolbox),
            radius: v => getDimensionUpdateValue(v, options, toolbox),
            radii: v => getRadiiUpdateValue(v, options, toolbox),
            spacings: v => getRadiiUpdateValue(v, options, toolbox),
        }, _ => {
            throw new Error('Unreachable: invalid tokens are supposed to be filter before');
        }, token.type, value);
        // It means that the dimension was an alias and has been updated before
        if (!output)
            return;
        tokenState.updateModeValue(mode, output);
    }));
}
