// @specifyapp/sdk/dist/httpApi/clients/createHttpApiClient.js
import { executeRepositoryRuleRouteDefinition, getMeRouteDefinition, getRepositoriesByOwnerRouteDefinition, getRepositoryDetailsByOwnerAndNameRouteDefinition, getRepositoryTokenTreeByOwnerAndNameRouteDefinition, } from '../routes/index.js';
import { createFetchClient } from './internals/createFetchClient.js';
export function createHttpApiClient(personalAccessToken) {
    return {
        getMe() {
            return createFetchClient(personalAccessToken).get({
                path: getMeRouteDefinition.makeRoute(),
            });
        },
        getRepositoriesByOwner(owner) {
            return createFetchClient(personalAccessToken).get({
                path: getRepositoriesByOwnerRouteDefinition.makeRoute({ owner }),
            });
        },
        getRepositoryByOwnerAndName(owner, name) {
            return createFetchClient(personalAccessToken).get({
                path: getRepositoryDetailsByOwnerAndNameRouteDefinition.makeRoute({ owner, name }),
            });
        },
        async getRepositoryTokenTreeByOwnerAndName(owner, name) {
            const { payload } = await createFetchClient(personalAccessToken).get({
                path: getRepositoryTokenTreeByOwnerAndNameRouteDefinition.makeRoute({ owner, name }),
            });
            return payload.tokenTree;
        },
        /**
         * @deprecated use `createRPCClient` instead
         * @param owner
         * @param name
         * @param rule
         */
        executeRepositoryRuleByOwnerAndName(owner, name, rule) {
            return createFetchClient(personalAccessToken).post({
                path: executeRepositoryRuleRouteDefinition.makeRoute({ owner, name }),
                body: rule,
            });
        },
    };
}
