// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/_JSON.js
import { z } from 'zod';
import { specifyModeAndValueLevelAliasSignatureSchema, } from '../internals/designTokenAlias.js';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyHexadecimalColorStringTypeName, specifyArcDegreeNumberTypeName, specifyIntegerNumberTypeName, specifyPercentageNumberTypeName, specifyPositiveIntegerNumberTypeName, specifyPositiveNumberTypeName, specifyRGBColorNumberTypeName, specifyZeroToOneNumberTypeName, specifyJSONStringTypeName, specifyJSONArrayTypeName, specifyJSONBooleanTypeName, specifyJSONNullTypeName, specifyJSONNumberTypeName, specifyJSONObjectTypeName, } from '../designTokenTypeNames.js';
/* ==========================================
   JSON Value Types
============================================= */
export const makeSpecifyJSONStringValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.string());
export const specifyJSONStringDefinition = createDesignTokenDefinition({
    type: specifyJSONStringTypeName,
    aliasableValueZodSchema: makeSpecifyJSONStringValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONStringValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyJSONStringTypeName },
            { _tokenType: specifyHexadecimalColorStringTypeName },
        ],
    },
});
export const makeSpecifyJSONNumberValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.number());
export const JSONNumberTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyJSONNumberTypeName },
        { _tokenType: specifyArcDegreeNumberTypeName },
        { _tokenType: specifyIntegerNumberTypeName },
        { _tokenType: specifyPercentageNumberTypeName },
        { _tokenType: specifyPositiveIntegerNumberTypeName },
        { _tokenType: specifyPositiveNumberTypeName },
        { _tokenType: specifyRGBColorNumberTypeName },
        { _tokenType: specifyZeroToOneNumberTypeName },
    ],
};
export const specifyJSONNumberDefinition = createDesignTokenDefinition({
    type: specifyJSONNumberTypeName,
    aliasableValueZodSchema: makeSpecifyJSONNumberValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONNumberValueSchema(false),
    tokenTypesMapping: JSONNumberTokenTypesMapping,
});
export const makeSpecifyJSONBooleanValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.boolean());
export const specifyJSONBooleanDefinition = createDesignTokenDefinition({
    type: specifyJSONBooleanTypeName,
    aliasableValueZodSchema: makeSpecifyJSONBooleanValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONBooleanValueSchema(false),
});
export const makeSpecifyJSONNullValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.null());
export const specifyJSONNullDefinition = createDesignTokenDefinition({
    type: specifyJSONNullTypeName,
    aliasableValueZodSchema: makeSpecifyJSONNullValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONNullValueSchema(false),
});
// The JSONLiteralValue schema definition MUST NOT be used in functional SDTF code.
// It is defined to be a primitive definition of this very module.
const makeJSONLiteralValueSchema = (isSupportingAliasing) => {
    return z.union([
        makeSpecifyJSONStringValueSchema(isSupportingAliasing),
        makeSpecifyJSONNumberValueSchema(isSupportingAliasing),
        makeSpecifyJSONBooleanValueSchema(isSupportingAliasing),
        makeSpecifyJSONNullValueSchema(isSupportingAliasing),
    ]);
};
// JSONLiteralValue is not aliasable since never implemented in SDTF
// The JSONValue schema definition MUST NOT be used in functional SDTF code.
// It is defined to be a primitive definition of this very module. Export is test only.
export const makeJSONValueSchema = (isSupportingAliasing) => z.lazy(() => z.union([
    makeJSONLiteralValueSchema(isSupportingAliasing),
    z.array(makeJSONValueSchema(isSupportingAliasing)),
    z.record(makeJSONValueSchema(isSupportingAliasing)),
]));
export function makeSpecifyJSONArrayValueSchema(isSupportingAliasing) {
    if (isSupportingAliasing) {
        return z.union([
            z.array(makeJSONValueSchema(isSupportingAliasing)),
            specifyModeAndValueLevelAliasSignatureSchema,
        ]);
    }
    return z.array(makeJSONValueSchema(isSupportingAliasing));
}
export const specifyJSONArrayDefinition = createDesignTokenDefinition({
    type: specifyJSONArrayTypeName,
    aliasableValueZodSchema: makeSpecifyJSONArrayValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONArrayValueSchema(false),
});
export function makeSpecifyJSONObjectValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.record(makeJSONValueSchema(isSupportingAliasing)))(isSupportingAliasing);
}
export const specifyJSONObjectDefinition = createDesignTokenDefinition({
    type: specifyJSONObjectTypeName,
    aliasableValueZodSchema: makeSpecifyJSONObjectValueSchema(true),
    resolvedValueZodSchema: makeSpecifyJSONObjectValueSchema(false),
});
