// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/bitmap.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyJSONStringValueSchema } from './_JSON.js';
import { makeSpecifyBitmapFormatValueSchema } from './bitmapFormat.js';
import { makeSpecifyPositiveIntegerNumberValueSchema } from './_numbers.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyBitmapTypeName, specifyBitmapFormatTypeName, specifyPositiveIntegerNumberTypeName, specifyJSONStringTypeName, } from '../designTokenTypeNames.js';
export const bitmapProviderValues = ['external', 'Specify'];
export const specifyBitmapProviderSchema = z.union([z.literal('external'), z.literal('Specify')]);
export function makeSpecifyBitmapValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        url: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
        format: makeSpecifyBitmapFormatValueSchema(isSupportingAliasing),
        width: makeSpecifyPositiveIntegerNumberValueSchema(isSupportingAliasing).nullable(),
        height: makeSpecifyPositiveIntegerNumberValueSchema(isSupportingAliasing).nullable(),
        variationLabel: makeSpecifyJSONStringValueSchema(isSupportingAliasing).nullable(),
        provider: specifyBitmapProviderSchema,
    })
        .strict())(isSupportingAliasing);
}
export const bitmapTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyBitmapTypeName },
        {
            _mapOf: {
                url: {
                    _tokenType: specifyJSONStringTypeName,
                },
                format: {
                    _tokenType: specifyBitmapFormatTypeName,
                },
                width: {
                    _tokenType: specifyPositiveIntegerNumberTypeName,
                },
                height: {
                    _tokenType: specifyPositiveIntegerNumberTypeName,
                },
                variationLabel: {
                    _tokenType: specifyJSONStringTypeName,
                },
                provider: {
                    _unionOf: bitmapProviderValues.map(provider => ({ _primitive: provider })),
                },
            },
        },
    ],
};
export const specifyBitmapDefinition = createDesignTokenDefinition({
    type: specifyBitmapTypeName,
    aliasableValueZodSchema: makeSpecifyBitmapValueSchema(true),
    resolvedValueZodSchema: makeSpecifyBitmapValueSchema(false),
    tokenTypesMapping: bitmapTokenTypesMapping,
});
