// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/bitmaps.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyBitmapValueSchema, bitmapTokenTypesMapping } from './bitmap.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyBitmapsTypeName } from '../designTokenTypeNames.js';
export function makeSpecifyBitmapsValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.object({
        files: z.array(makeSpecifyBitmapValueSchema(isSupportingAliasing)).nonempty(),
    }))(isSupportingAliasing);
}
export const specifyBitmapsDefinition = createDesignTokenDefinition({
    type: specifyBitmapsTypeName,
    aliasableValueZodSchema: makeSpecifyBitmapsValueSchema(true),
    resolvedValueZodSchema: makeSpecifyBitmapsValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyBitmapsTypeName },
            { _mapOf: { files: { _arrayOf: [bitmapTokenTypesMapping] } } },
        ],
    },
});
