// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/blur.js
import { makeSpecifyDimensionValueSchema } from './dimension.js';
import { specifyBlurTypeName, specifyDimensionTypeName, specifyDimensionUnitTypeName, } from '../designTokenTypeNames.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { JSONNumberTokenTypesMapping } from './_JSON.js';
export const makeSpecifyBlurValueSchema = makeSpecifyDimensionValueSchema;
export const specifyBlurDefinition = createDesignTokenDefinition({
    type: specifyBlurTypeName,
    aliasableValueZodSchema: makeSpecifyBlurValueSchema(true),
    resolvedValueZodSchema: makeSpecifyBlurValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyBlurTypeName },
            { _tokenType: specifyDimensionTypeName },
            {
                _mapOf: {
                    unit: {
                        _tokenType: specifyDimensionUnitTypeName,
                    },
                    value: JSONNumberTokenTypesMapping,
                },
            },
        ],
    },
});
