// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/borderStyle.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyBorderStyleLineCapValueSchema } from './borderStyleLineCap.js';
import { makeSpecifyDimensionValueSchema } from './dimension.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyBorderStyleTypeName } from '../designTokenTypeNames.js';
export function makeSpecifyObjectBorderStyleValueSchema(isSupportingAliasing) {
    return z
        .object({
        dashArray: z.array(makeSpecifyDimensionValueSchema(isSupportingAliasing)),
        lineCap: makeSpecifyBorderStyleLineCapValueSchema(isSupportingAliasing),
    })
        .strict();
}
// Named Border Style
// The SpecifyNamedBorderStyleValue schema definition MUST NOT be used in functional SDTF code.
// It is defined to be a primitive definition of this very module. Export is test only.
export const specifyNamedBorderStyleValues = [
    'none',
    'hidden',
    'dotted',
    'dashed',
    'solid',
    'double',
    'groove',
    'ridge',
    'inset',
    'outset',
    'inherit',
    'initial',
    'unset',
];
export const makeSpecifyNamedBorderStyleValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.union([
    z.literal('none'),
    z.literal('hidden'),
    z.literal('dotted'),
    z.literal('dashed'),
    z.literal('solid'),
    z.literal('double'),
    z.literal('groove'),
    z.literal('ridge'),
    z.literal('inset'),
    z.literal('outset'),
    z.literal('inherit'),
    z.literal('initial'),
    z.literal('unset'),
]));
export function makeSpecifyBorderStyleValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.union([
        makeSpecifyObjectBorderStyleValueSchema(isSupportingAliasing),
        makeSpecifyNamedBorderStyleValueSchema(isSupportingAliasing),
    ]))(isSupportingAliasing);
}
export const borderStyleDefinition = createDesignTokenDefinition({
    type: specifyBorderStyleTypeName,
    aliasableValueZodSchema: makeSpecifyBorderStyleValueSchema(true),
    resolvedValueZodSchema: makeSpecifyBorderStyleValueSchema(false),
});
