// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/border.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyBorderStyleValueSchema } from './borderStyle.js';
import { makeSpecifyColorValueSchema } from './color.js';
import { makeSpecifyDimensionValueSchema } from './dimension.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyBorderStyleTypeName, specifyBorderTypeName, specifyColorTypeName, specifyDimensionTypeName, } from '../designTokenTypeNames.js';
import { radiiTokenTypesMapping } from './radii.js';
export function makeSpecifyBorderValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        color: makeSpecifyColorValueSchema(isSupportingAliasing),
        style: makeSpecifyBorderStyleValueSchema(isSupportingAliasing),
        width: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        rectangleCornerRadii: z.union([
            z.tuple([
                makeSpecifyDimensionValueSchema(isSupportingAliasing),
                makeSpecifyDimensionValueSchema(isSupportingAliasing),
                makeSpecifyDimensionValueSchema(isSupportingAliasing),
                makeSpecifyDimensionValueSchema(isSupportingAliasing),
            ]),
            z.null(),
        ]),
    })
        .strict())(isSupportingAliasing);
}
export const specifyBorderDefinition = createDesignTokenDefinition({
    type: specifyBorderTypeName,
    aliasableValueZodSchema: makeSpecifyBorderValueSchema(true),
    resolvedValueZodSchema: makeSpecifyBorderValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyBorderTypeName },
            {
                _mapOf: {
                    color: {
                        _tokenType: specifyColorTypeName,
                    },
                    style: {
                        _tokenType: specifyBorderStyleTypeName,
                    },
                    width: {
                        _tokenType: specifyDimensionTypeName,
                    },
                    rectangleCornerRadii: radiiTokenTypesMapping,
                },
            },
        ],
    },
});
