// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/color.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyHexadecimalColorStringValueSchema } from './_strings.js';
import { makeSpecifyArcDegreeNumberValueSchema, makeSpecifyPercentageNumberValueSchema, makeSpecifyPositiveNumberValueSchema, makeSpecifyRGBColorNumberValueSchema, makeSpecifyZeroToOneNumberValueSchema, } from './_numbers.js';
import { makeSpecifyJSONNumberValueSchema } from './_JSON.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyArcDegreeNumberTypeName, specifyColorTypeName, specifyHexadecimalColorStringTypeName, specifyJSONNumberTypeName, specifyOpacityTypeName, specifyPercentageNumberTypeName, specifyPositiveNumberTypeName, specifyRGBColorNumberTypeName, specifyZeroToOneNumberTypeName, } from '../designTokenTypeNames.js';
export const specifyColorModelNames = ['hex', 'rgb', 'hsl', 'hsb', 'lch', 'lab'];
export function makeSpecifyColorValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.discriminatedUnion('model', [
        z
            .object({
            model: z.literal('hex'),
            hex: makeSpecifyHexadecimalColorStringValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
        z
            .object({
            model: z.literal('rgb'),
            red: makeSpecifyRGBColorNumberValueSchema(isSupportingAliasing),
            green: makeSpecifyRGBColorNumberValueSchema(isSupportingAliasing),
            blue: makeSpecifyRGBColorNumberValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
        z
            .object({
            model: z.literal('hsl'),
            hue: makeSpecifyArcDegreeNumberValueSchema(isSupportingAliasing),
            saturation: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            lightness: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
        z
            .object({
            model: z.literal('hsb'),
            hue: makeSpecifyArcDegreeNumberValueSchema(isSupportingAliasing),
            saturation: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            brightness: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
        z
            .object({
            model: z.literal('lch'),
            lightness: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            chroma: makeSpecifyPositiveNumberValueSchema(isSupportingAliasing),
            hue: makeSpecifyArcDegreeNumberValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
        z
            .object({
            model: z.literal('lab'),
            lightness: makeSpecifyPercentageNumberValueSchema(isSupportingAliasing),
            aAxis: makeSpecifyJSONNumberValueSchema(isSupportingAliasing),
            bAxis: makeSpecifyJSONNumberValueSchema(isSupportingAliasing),
            alpha: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
        })
            .strict(),
    ]))(isSupportingAliasing);
}
export const colorTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyColorTypeName },
        {
            _discriminator: 'model',
            _discriminatedUnionOf: [
                {
                    _mapOf: {
                        model: { _primitive: 'hex' },
                        hex: { _tokenType: specifyHexadecimalColorStringTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        model: { _primitive: 'rgb' },
                        red: { _tokenType: specifyRGBColorNumberTypeName },
                        green: { _tokenType: specifyRGBColorNumberTypeName },
                        blue: { _tokenType: specifyRGBColorNumberTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        model: { _primitive: 'hsl' },
                        hue: { _tokenType: specifyArcDegreeNumberTypeName },
                        saturation: { _tokenType: specifyPercentageNumberTypeName },
                        lightness: { _tokenType: specifyPercentageNumberTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        model: { _primitive: 'hsb' },
                        hue: { _tokenType: specifyArcDegreeNumberTypeName },
                        saturation: { _tokenType: specifyPercentageNumberTypeName },
                        brightness: { _tokenType: specifyPercentageNumberTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        model: { _primitive: 'lch' },
                        lightness: { _tokenType: specifyPercentageNumberTypeName },
                        chroma: { _tokenType: specifyPositiveNumberTypeName },
                        hue: { _tokenType: specifyArcDegreeNumberTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        model: { _primitive: 'lab' },
                        lightness: { _tokenType: specifyPercentageNumberTypeName },
                        aAxis: { _tokenType: specifyJSONNumberTypeName },
                        bAxis: { _tokenType: specifyJSONNumberTypeName },
                        alpha: {
                            _unionOf: [
                                { _tokenType: specifyZeroToOneNumberTypeName },
                                { _tokenType: specifyOpacityTypeName },
                            ],
                        },
                    },
                },
            ],
        },
    ],
};
export const specifyColorDefinition = createDesignTokenDefinition({
    type: specifyColorTypeName,
    aliasableValueZodSchema: makeSpecifyColorValueSchema(true),
    resolvedValueZodSchema: makeSpecifyColorValueSchema(false),
    tokenTypesMapping: colorTokenTypesMapping,
});
