// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/dimension.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { JSONNumberTokenTypesMapping, makeSpecifyJSONNumberValueSchema } from './_JSON.js';
import { makeSpecifyDimensionUnitValueSchema } from './dimensionUnit.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyDimensionTypeName, specifyDimensionUnitTypeName } from '../designTokenTypeNames.js';
export function makeSpecifyDimensionValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        value: makeSpecifyJSONNumberValueSchema(isSupportingAliasing),
        unit: makeSpecifyDimensionUnitValueSchema(isSupportingAliasing),
    })
        .strict())(isSupportingAliasing);
}
export const specifyDimensionTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyDimensionTypeName },
        {
            _mapOf: {
                unit: {
                    _tokenType: specifyDimensionUnitTypeName,
                },
                value: JSONNumberTokenTypesMapping,
            },
        },
    ],
};
export const specifyDimensionDefinition = createDesignTokenDefinition({
    type: specifyDimensionTypeName,
    aliasableValueZodSchema: makeSpecifyDimensionValueSchema(true),
    resolvedValueZodSchema: makeSpecifyDimensionValueSchema(false),
    tokenTypesMapping: specifyDimensionTokenTypesMapping,
});
