// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/duration.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyDurationUnitValueSchema } from './durationUnit.js';
import { makeSpecifyPositiveNumberValueSchema } from './_numbers.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyDurationTypeName, specifyDurationUnitTypeName, specifyPositiveNumberTypeName, } from '../designTokenTypeNames.js';
export function makeSpecifyDurationValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        value: makeSpecifyPositiveNumberValueSchema(isSupportingAliasing),
        unit: makeSpecifyDurationUnitValueSchema(isSupportingAliasing),
    })
        .strict())(isSupportingAliasing);
}
export const specifyDurationDefinition = createDesignTokenDefinition({
    type: specifyDurationTypeName,
    aliasableValueZodSchema: makeSpecifyDurationValueSchema(true),
    resolvedValueZodSchema: makeSpecifyDurationValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyDurationTypeName },
            {
                _mapOf: {
                    value: {
                        _tokenType: specifyPositiveNumberTypeName,
                    },
                    unit: {
                        _tokenType: specifyDurationUnitTypeName,
                    },
                },
            },
        ],
    },
});
