// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/fontFeatures.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyFontFeatureValueSchema } from './fontFeature.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyFontFeaturesTypeName, specifyFontFeatureTypeName, } from '../designTokenTypeNames.js';
export function makeSpecifyFontFeaturesValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.array(makeSpecifyFontFeatureValueSchema(isSupportingAliasing)).nonempty())(isSupportingAliasing);
}
export const specifyFontFeaturesDefinition = createDesignTokenDefinition({
    type: specifyFontFeaturesTypeName,
    aliasableValueZodSchema: makeSpecifyFontFeaturesValueSchema(true),
    resolvedValueZodSchema: makeSpecifyFontFeaturesValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyFontFeaturesTypeName },
            { _arrayOf: [{ _tokenType: specifyFontFeatureTypeName }] },
        ],
    },
});
