// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/font.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyFontFamilyValueSchema } from './fontFamily.js';
import { makeSpecifyJSONStringValueSchema } from './_JSON.js';
import { makeSpecifyFontWeightValueSchema } from './fontWeight.js';
import { makeSpecifyFontStyleValueSchema } from './fontStyle.js';
import { makeSpecifyFontFormatValueSchema } from './fontFormat.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyFontFamilyTypeName, specifyFontFormatTypeName, specifyFontStyleTypeName, specifyFontTypeName, specifyFontWeightTypeName, specifyJSONStringTypeName, } from '../designTokenTypeNames.js';
export const specifyFontProviderValues = [
    'external',
    'Specify',
    'Google Fonts',
    'Adobe Fonts',
];
export const specifyFontProviderSchema = z.union([
    z.literal('external'),
    z.literal('Specify'),
    z.literal('Google Fonts'),
    z.literal('Adobe Fonts'),
]);
export function makeSpecifyFontValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        family: makeSpecifyFontFamilyValueSchema(isSupportingAliasing),
        postScriptName: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
        weight: makeSpecifyFontWeightValueSchema(isSupportingAliasing),
        style: makeSpecifyFontStyleValueSchema(isSupportingAliasing),
        files: z.array(z
            .object({
            url: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
            format: makeSpecifyFontFormatValueSchema(isSupportingAliasing),
            provider: specifyFontProviderSchema,
        })
            .strict()),
    })
        .strict())(isSupportingAliasing);
}
export const fontTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyFontTypeName },
        {
            _mapOf: {
                family: {
                    _tokenType: specifyFontFamilyTypeName,
                },
                postScriptName: {
                    _tokenType: specifyJSONStringTypeName,
                },
                weight: {
                    _tokenType: specifyFontWeightTypeName,
                },
                style: {
                    _tokenType: specifyFontStyleTypeName,
                },
                files: {
                    _arrayOf: [
                        {
                            _mapOf: {
                                url: { _tokenType: specifyJSONStringTypeName },
                                format: { _tokenType: specifyFontFormatTypeName },
                                provider: {
                                    _unionOf: specifyFontProviderValues.map(provider => ({ _primitive: provider })),
                                },
                            },
                        },
                    ],
                },
            },
        },
    ],
};
export const specifyFontDefinition = createDesignTokenDefinition({
    type: specifyFontTypeName,
    aliasableValueZodSchema: makeSpecifyFontValueSchema(true),
    resolvedValueZodSchema: makeSpecifyFontValueSchema(false),
    tokenTypesMapping: fontTokenTypesMapping,
});
