// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/gradient.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyArcDegreeNumberValueSchema, makeSpecifyZeroToOneNumberValueSchema, } from './_numbers.js';
import { colorTokenTypesMapping, makeSpecifyColorValueSchema } from './color.js';
import { makeSpecifyJSONStringValueSchema } from './_JSON.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyArcDegreeNumberTypeName, specifyGradientTypeName, specifyJSONStringTypeName, specifyZeroToOneNumberTypeName, } from '../designTokenTypeNames.js';
export const specifyGradientTypeValues = ['linear', 'radial', 'conic'];
export function makeSpecifyGradientValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.discriminatedUnion('type', [
        z
            .object({
            type: z.literal('linear'),
            angle: makeSpecifyArcDegreeNumberValueSchema(isSupportingAliasing),
            colorStops: z
                .array(z
                .object({
                color: makeSpecifyColorValueSchema(true),
                position: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
            })
                .strict())
                .nonempty(),
        })
            .strict(),
        z
            .object({
            type: z.literal('radial'),
            position: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
            colorStops: z
                .array(z
                .object({
                color: makeSpecifyColorValueSchema(true),
                position: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
            })
                .strict())
                .nonempty(),
        })
            .strict(),
        z
            .object({
            type: z.literal('conic'),
            angle: makeSpecifyArcDegreeNumberValueSchema(isSupportingAliasing),
            position: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
            colorStops: z
                .array(z
                .object({
                color: makeSpecifyColorValueSchema(true),
                position: makeSpecifyZeroToOneNumberValueSchema(isSupportingAliasing),
            })
                .strict())
                .nonempty(),
        })
            .strict(),
    ]))(isSupportingAliasing);
}
export const gradientTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyGradientTypeName },
        {
            _discriminator: 'type',
            _discriminatedUnionOf: [
                {
                    _mapOf: {
                        type: { _primitive: 'linear' },
                        angle: { _tokenType: specifyArcDegreeNumberTypeName },
                        colorStops: {
                            _arrayOf: [
                                {
                                    _mapOf: {
                                        color: colorTokenTypesMapping,
                                        position: { _tokenType: specifyZeroToOneNumberTypeName },
                                    },
                                },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        type: { _primitive: 'radial' },
                        position: { _tokenType: specifyJSONStringTypeName },
                        colorStops: {
                            _arrayOf: [
                                {
                                    _mapOf: {
                                        color: colorTokenTypesMapping,
                                        position: { _tokenType: specifyZeroToOneNumberTypeName },
                                    },
                                },
                            ],
                        },
                    },
                },
                {
                    _mapOf: {
                        type: { _primitive: 'conic' },
                        angle: { _tokenType: specifyArcDegreeNumberTypeName },
                        position: { _tokenType: specifyJSONStringTypeName },
                        colorStops: {
                            _arrayOf: [
                                {
                                    _mapOf: {
                                        color: colorTokenTypesMapping,
                                        position: { _tokenType: specifyZeroToOneNumberTypeName },
                                    },
                                },
                            ],
                        },
                    },
                },
            ],
        },
    ],
};
// This schema has been created specicically for the update
// The color validation being handled on his own side, we don't care about validating it here
export const specifyGradientPartialAlisableValueColorStops = z
    .array(z
    .object({
    position: makeSpecifyZeroToOneNumberValueSchema(true),
})
    .passthrough()
    .partial())
    .nonempty();
export const specifyGradientDefinition = createDesignTokenDefinition({
    type: specifyGradientTypeName,
    aliasableValueZodSchema: makeSpecifyGradientValueSchema(true),
    resolvedValueZodSchema: makeSpecifyGradientValueSchema(false),
    tokenTypesMapping: gradientTokenTypesMapping,
});
