// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/gradients.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { gradientTokenTypesMapping, makeSpecifyGradientValueSchema } from './gradient.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyGradientsTypeName } from '../designTokenTypeNames.js';
export function makeSpecifyGradientsValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.array(makeSpecifyGradientValueSchema(isSupportingAliasing)).nonempty())(isSupportingAliasing);
}
export const specifyGradientsDefinition = createDesignTokenDefinition({
    type: specifyGradientsTypeName,
    aliasableValueZodSchema: makeSpecifyGradientsValueSchema(true),
    resolvedValueZodSchema: makeSpecifyGradientsValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [{ _tokenType: specifyGradientsTypeName }, { _arrayOf: [gradientTokenTypesMapping] }],
    },
});
