// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/radius.js
import { makeSpecifyDimensionValueSchema, } from './dimension.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyDimensionTypeName, specifyDimensionUnitTypeName, specifyRadiusTypeName, } from '../designTokenTypeNames.js';
import { JSONNumberTokenTypesMapping } from './_JSON.js';
export function makeSpecifyRadiusValueSchema(isSupportingAliasing) {
    return makeSpecifyDimensionValueSchema(isSupportingAliasing);
}
export const specifyRadiusDefinition = createDesignTokenDefinition({
    type: specifyRadiusTypeName,
    aliasableValueZodSchema: makeSpecifyRadiusValueSchema(true),
    resolvedValueZodSchema: makeSpecifyRadiusValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyRadiusTypeName },
            { _tokenType: specifyDimensionTypeName },
            {
                _mapOf: {
                    unit: {
                        _tokenType: specifyDimensionUnitTypeName,
                    },
                    value: JSONNumberTokenTypesMapping,
                },
            },
        ],
    },
});
