// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/radii.js
import { z } from 'zod';
import { makeSpecifyRadiusValueSchema } from './radius.js';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { specifyRadiiTypeName, specifyRadiusTypeName } from '../designTokenTypeNames.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
export function makeSpecifyRadiiValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(
    // We make the implementation more performant by using a single z.array instead of z.union of z.tuple
    z
        .array(makeSpecifyRadiusValueSchema(isSupportingAliasing), {
        required_error: 'Specify radii must be an array containing 1 to 4 spacing value(s)',
        invalid_type_error: 'Specify radii must be an array containing 1 to 4 spacing value(s)',
    })
        .min(1, { message: 'Specify radii must have at least 1 radius value(s)' })
        .max(4, { message: 'Specify radii must have at most 4 radius values' }))(isSupportingAliasing);
}
export const radiiTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyRadiiTypeName },
        { _tuple: [{ _tokenType: specifyRadiusTypeName }] },
        { _tuple: [{ _tokenType: specifyRadiusTypeName }, { _tokenType: specifyRadiusTypeName }] },
        {
            _tuple: [
                { _tokenType: specifyRadiusTypeName },
                { _tokenType: specifyRadiusTypeName },
                { _tokenType: specifyRadiusTypeName },
            ],
        },
        {
            _tuple: [
                { _tokenType: specifyRadiusTypeName },
                { _tokenType: specifyRadiusTypeName },
                { _tokenType: specifyRadiusTypeName },
                { _tokenType: specifyRadiusTypeName },
            ],
        },
    ],
};
export const specifyRadiiDefinition = createDesignTokenDefinition({
    type: specifyRadiiTypeName,
    aliasableValueZodSchema: makeSpecifyRadiiValueSchema(true),
    resolvedValueZodSchema: makeSpecifyRadiiValueSchema(false),
    tokenTypesMapping: radiiTokenTypesMapping,
});
