// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/shadow.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { colorTokenTypesMapping, makeSpecifyColorValueSchema } from './color.js';
import { makeSpecifyDimensionValueSchema } from './dimension.js';
import { makeSpecifyShadowTypeValueSchema } from './shadowType.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyBlurTypeName, specifyDimensionTypeName, specifyRadiusTypeName, specifyShadowTypeName, specifyShadowTypeTypeName, } from '../designTokenTypeNames.js';
export function makeSpecifyShadowValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        color: makeSpecifyColorValueSchema(isSupportingAliasing),
        offsetX: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        offsetY: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        blurRadius: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        spreadRadius: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        type: makeSpecifyShadowTypeValueSchema(isSupportingAliasing),
    })
        .strict())(isSupportingAliasing);
}
export const shadowTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyShadowTypeName },
        {
            _mapOf: {
                color: colorTokenTypesMapping,
                offsetX: { _tokenType: specifyDimensionTypeName },
                offsetY: { _tokenType: specifyDimensionTypeName },
                blurRadius: {
                    _unionOf: [
                        { _tokenType: specifyBlurTypeName },
                        { _tokenType: specifyRadiusTypeName },
                        { _tokenType: specifyDimensionTypeName },
                    ],
                },
                spreadRadius: {
                    _unionOf: [
                        { _tokenType: specifyBlurTypeName },
                        { _tokenType: specifyRadiusTypeName },
                        { _tokenType: specifyDimensionTypeName },
                    ],
                },
                type: { _tokenType: specifyShadowTypeTypeName },
            },
        },
    ],
};
export const specifyShadowDefinition = createDesignTokenDefinition({
    type: specifyShadowTypeName,
    aliasableValueZodSchema: makeSpecifyShadowValueSchema(true),
    resolvedValueZodSchema: makeSpecifyShadowValueSchema(false),
    tokenTypesMapping: shadowTokenTypesMapping,
});
