// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/spacings.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifySpacingsTypeName, specifySpacingTypeName } from '../designTokenTypeNames.js';
import { makeSpecifySpacingValueSchema } from './spacing.js';
export function makeSpecifySpacingsValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(
    // We make the implementation more performant by using a single z.array instead of z.union of z.tuple
    z
        .array(makeSpecifySpacingValueSchema(isSupportingAliasing), {
        required_error: 'Specify spacings must be an array containing 1 to 4 spacing value(s)',
        invalid_type_error: 'Specify radii must be an array containing 1 to 4 spacing value(s)',
    })
        .min(1, { message: 'Specify spacings must have at least 1 spacing value(s)' })
        .max(4, { message: 'Specify spacings must have at most 4 spacing values' }))(isSupportingAliasing);
}
export const specifySpacingsDefinition = createDesignTokenDefinition({
    type: specifySpacingsTypeName,
    aliasableValueZodSchema: makeSpecifySpacingsValueSchema(true),
    resolvedValueZodSchema: makeSpecifySpacingsValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifySpacingsTypeName },
            {
                _tuple: [{ _tokenType: specifySpacingTypeName }],
            },
            {
                _tuple: [{ _tokenType: specifySpacingTypeName }, { _tokenType: specifySpacingTypeName }],
            },
            {
                _tuple: [
                    { _tokenType: specifySpacingTypeName },
                    { _tokenType: specifySpacingTypeName },
                    { _tokenType: specifySpacingTypeName },
                ],
            },
            {
                _tuple: [
                    { _tokenType: specifySpacingTypeName },
                    { _tokenType: specifySpacingTypeName },
                    { _tokenType: specifySpacingTypeName },
                    { _tokenType: specifySpacingTypeName },
                ],
            },
        ],
    },
});
