// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/stepsTimingFunction.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyPositiveIntegerNumberValueSchema } from './_numbers.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyPositiveIntegerNumberTypeName, specifyStepsTimingFunctionTypeName, } from '../designTokenTypeNames.js';
export const specifyStepsTimingFunctionJumpTermValues = [
    'start',
    'end',
    'jump-start',
    'jump-end',
    'jump-none',
    'jump-both',
];
const stepsTimingFunctionJumpTermSchema = z.union([
    z.literal('start'),
    z.literal('end'),
    z.literal('jump-start'),
    z.literal('jump-end'),
    z.literal('jump-none'),
    z.literal('jump-both'),
]);
export function makeSpecifyStepsTimingFunctionValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        stepsCount: makeSpecifyPositiveIntegerNumberValueSchema(isSupportingAliasing),
        jumpTerm: stepsTimingFunctionJumpTermSchema,
    })
        .strict())(isSupportingAliasing);
}
export const specifyStepsTimingFunctionDefinition = createDesignTokenDefinition({
    type: specifyStepsTimingFunctionTypeName,
    aliasableValueZodSchema: makeSpecifyStepsTimingFunctionValueSchema(true),
    resolvedValueZodSchema: makeSpecifyStepsTimingFunctionValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyStepsTimingFunctionTypeName },
            {
                _mapOf: {
                    stepsCount: { _tokenType: specifyPositiveIntegerNumberTypeName },
                    jumpTerm: {
                        _unionOf: specifyStepsTimingFunctionJumpTermValues.map(value => ({
                            _primitive: value,
                        })),
                    },
                },
            },
        ],
    },
});
