// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/textAlignHorizontal.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyTextAlignHorizontalTypeName } from '../designTokenTypeNames.js';
export const specifyTextAlignHorizontalValues = [
    'initial',
    'left',
    'right',
    'center',
    'justify',
    'start',
    'end',
    'justify-all',
    'match-parent',
];
export const makeSpecifyTextAlignHorizontalValueSchema = makeUnionWithModeAndValueLevelAliasValue(z.union([
    z.literal('initial'),
    z.literal('left'),
    z.literal('right'),
    z.literal('center'),
    z.literal('justify'),
    z.literal('start'),
    z.literal('end'),
    z.literal('justify-all'),
    z.literal('match-parent'),
]));
export const specifyTextAlignHorizontalDefinition = createDesignTokenDefinition({
    type: specifyTextAlignHorizontalTypeName,
    aliasableValueZodSchema: makeSpecifyTextAlignHorizontalValueSchema(true),
    resolvedValueZodSchema: makeSpecifyTextAlignHorizontalValueSchema(false),
});
