// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/textStyle.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { fontTokenTypesMapping, makeSpecifyFontValueSchema } from './font.js';
import { specifyDimensionTokenTypesMapping, makeSpecifyDimensionValueSchema } from './dimension.js';
import { colorTokenTypesMapping, makeSpecifyColorValueSchema } from './color.js';
import { makeSpecifyFontFeaturesValueSchema } from './fontFeatures.js';
import { makeSpecifyTextAlignHorizontalValueSchema } from './textAlignHorizontal.js';
import { makeSpecifyTextAlignVerticalValueSchema } from './textAlignVertical.js';
import { makeSpecifyTextDecorationValueSchema } from './textDecoration.js';
import { makeSpecifyTextTransformValueSchema } from './textTransform.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyDimensionTypeName, specifyFontFeaturesTypeName, specifyFontFeatureTypeName, specifySpacingTypeName, specifyTextAlignHorizontalTypeName, specifyTextAlignVerticalTypeName, specifyTextDecorationTypeName, specifyTextStyleTypeName, specifyTextTransformTypeName, } from '../designTokenTypeNames.js';
import { specifySpacingTokenTypesMapping } from './spacing.js';
export function makeSpecifyTextStyleValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        font: makeSpecifyFontValueSchema(isSupportingAliasing),
        fontSize: makeSpecifyDimensionValueSchema(isSupportingAliasing),
        color: makeSpecifyColorValueSchema(isSupportingAliasing).nullable(),
        fontFeatures: makeSpecifyFontFeaturesValueSchema(isSupportingAliasing).nullable(),
        lineHeight: makeSpecifyDimensionValueSchema(isSupportingAliasing).nullable(),
        letterSpacing: makeSpecifyDimensionValueSchema(isSupportingAliasing).nullable(),
        paragraphSpacing: makeSpecifyDimensionValueSchema(isSupportingAliasing).nullable(),
        textAlignHorizontal: makeSpecifyTextAlignHorizontalValueSchema(isSupportingAliasing).nullable(),
        textAlignVertical: makeSpecifyTextAlignVerticalValueSchema(isSupportingAliasing).nullable(),
        textDecoration: makeSpecifyTextDecorationValueSchema(isSupportingAliasing).nullable(),
        textIndent: makeSpecifyDimensionValueSchema(isSupportingAliasing).nullable(),
        textTransform: makeSpecifyTextTransformValueSchema(isSupportingAliasing).nullable(),
    })
        .strict())(isSupportingAliasing);
}
export const specifyTextStyleDefinition = createDesignTokenDefinition({
    type: specifyTextStyleTypeName,
    aliasableValueZodSchema: makeSpecifyTextStyleValueSchema(true),
    resolvedValueZodSchema: makeSpecifyTextStyleValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyTextStyleTypeName },
            {
                _mapOf: {
                    font: fontTokenTypesMapping,
                    fontSize: specifyDimensionTokenTypesMapping,
                    color: colorTokenTypesMapping,
                    fontFeatures: {
                        _unionOf: [
                            { _tokenType: specifyFontFeaturesTypeName },
                            { _arrayOf: [{ _tokenType: specifyFontFeatureTypeName }] },
                        ],
                    },
                    lineHeight: { _tokenType: specifyDimensionTypeName },
                    letterSpacing: {
                        _unionOf: [
                            ...specifyDimensionTokenTypesMapping._unionOf,
                            ...specifySpacingTokenTypesMapping._unionOf,
                        ],
                    },
                    paragraphSpacing: {
                        _unionOf: [
                            { _tokenType: specifyDimensionTypeName },
                            { _tokenType: specifySpacingTypeName },
                        ],
                    },
                    textAlignHorizontal: { _tokenType: specifyTextAlignHorizontalTypeName },
                    textAlignVertical: { _tokenType: specifyTextAlignVerticalTypeName },
                    textDecoration: { _tokenType: specifyTextDecorationTypeName },
                    textIndent: {
                        _unionOf: [
                            ...specifyDimensionTokenTypesMapping._unionOf,
                            ...specifySpacingTokenTypesMapping._unionOf,
                        ],
                    },
                    textTransform: { _tokenType: specifyTextTransformTypeName },
                },
            },
        ],
    },
});
