// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/transition.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyDurationValueSchema } from './duration.js';
import { makeSpecifyCubicBezierValueSchema } from './cubicBezier.js';
import { makeSpecifyStepsTimingFunctionValueSchema } from './stepsTimingFunction.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyCubicBezierTypeName, specifyDurationTypeName, specifyStepsTimingFunctionTypeName, specifyTransitionTypeName, } from '../designTokenTypeNames.js';
export function makeSpecifyTransitionValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        duration: makeSpecifyDurationValueSchema(isSupportingAliasing),
        delay: makeSpecifyDurationValueSchema(isSupportingAliasing),
        timingFunction: z.union([
            makeSpecifyCubicBezierValueSchema(isSupportingAliasing),
            makeSpecifyStepsTimingFunctionValueSchema(isSupportingAliasing),
        ]),
    })
        .strict())(isSupportingAliasing);
}
export const specifyTransitionDefinition = createDesignTokenDefinition({
    type: specifyTransitionTypeName,
    aliasableValueZodSchema: makeSpecifyTransitionValueSchema(true),
    resolvedValueZodSchema: makeSpecifyTransitionValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyTransitionTypeName },
            {
                _mapOf: {
                    duration: { _tokenType: specifyDurationTypeName },
                    delay: { _tokenType: specifyDurationTypeName },
                    timingFunction: {
                        _unionOf: [
                            { _tokenType: specifyCubicBezierTypeName },
                            { _tokenType: specifyStepsTimingFunctionTypeName },
                        ],
                    },
                },
            },
        ],
    },
});
