// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/vector.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyJSONStringValueSchema } from './_JSON.js';
import { makeSpecifyVectorFormatValueSchema } from './vectorFormat.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyJSONStringTypeName, specifyVectorFormatTypeName, specifyVectorTypeName, } from '../designTokenTypeNames.js';
export const vectorProviderValues = ['external', 'Specify'];
export const specifyVectorProviderSchema = z.union([z.literal('external'), z.literal('Specify')]);
export function makeSpecifyVectorValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z
        .object({
        url: makeSpecifyJSONStringValueSchema(isSupportingAliasing),
        format: makeSpecifyVectorFormatValueSchema(isSupportingAliasing),
        variationLabel: makeSpecifyJSONStringValueSchema(isSupportingAliasing).nullable(),
        provider: specifyVectorProviderSchema,
    })
        .strict())(isSupportingAliasing);
}
export const vectorTokenTypesMapping = {
    _unionOf: [
        { _tokenType: specifyVectorTypeName },
        {
            _mapOf: {
                url: { _tokenType: specifyJSONStringTypeName },
                format: { _tokenType: specifyVectorFormatTypeName },
                variationLabel: { _tokenType: specifyJSONStringTypeName },
                provider: {
                    _unionOf: vectorProviderValues.map(provider => ({ _primitive: provider })),
                },
            },
        },
    ],
};
export const specifyVectorDefinition = createDesignTokenDefinition({
    type: specifyVectorTypeName,
    aliasableValueZodSchema: makeSpecifyVectorValueSchema(true),
    resolvedValueZodSchema: makeSpecifyVectorValueSchema(false),
    tokenTypesMapping: vectorTokenTypesMapping,
});
