// @specifyapp/specify-design-token-format/dist/definitions/tokenTypes/vectors.js
import { z } from 'zod';
import { makeUnionWithModeAndValueLevelAliasValue } from '../internals/makeUnionWithModeAndValueLevelAliasValue.js';
import { makeSpecifyVectorValueSchema, vectorTokenTypesMapping } from './vector.js';
import { createDesignTokenDefinition } from '../internals/createDesignTokenDefinition.js';
import { specifyVectorsTypeName } from '../designTokenTypeNames.js';
export function makeSpecifyVectorsValueSchema(isSupportingAliasing) {
    return makeUnionWithModeAndValueLevelAliasValue(z.object({ files: z.array(makeSpecifyVectorValueSchema(isSupportingAliasing)).nonempty() }))(isSupportingAliasing);
}
export const specifyVectorsDefinition = createDesignTokenDefinition({
    type: specifyVectorsTypeName,
    aliasableValueZodSchema: makeSpecifyVectorsValueSchema(true),
    resolvedValueZodSchema: makeSpecifyVectorsValueSchema(false),
    tokenTypesMapping: {
        _unionOf: [
            { _tokenType: specifyVectorsTypeName },
            { _mapOf: { files: { _arrayOf: [vectorTokenTypesMapping] } } },
        ],
    },
});
