// @specifyapp/sdk/dist/parsersEngine/internals/makeRemoteParsersPipelinesFromRules.js
import { getBuiltInParserDefinition } from '../../builtInParsers/index.js';
import { pickLastParserInParserRule } from '../../builtInParsers/utils/pickLastParserInParserRule.js';
import { createRPCClient } from '../../httpApi/clients/createRPCClient.js';
import { handleSerializedParsersEngineResult } from './handleSerializedParsersEngineResult.js';
import { SpecifyError, specifyErrors } from '../../errors/index.js';
import { serializeParsersEngineDataBox } from '../utils/serializeParsersEngineDataBox.js';
/**
 * Build a list of ParserFunctions for HTTP RPC execution wrapping all Rules to use a single HTTP request
 * @param rules
 * @param context
 */
export function makeRemoteParsersPipelinesFromRules(rules, context) {
    let futureResults;
    function getOrFetchResults(rules, dataBox) {
        if (futureResults) {
            return futureResults;
        }
        futureResults = createRPCClient(context.personalAccessToken).executeParsersEngine({
            body: {
                rules,
                dataBox,
                returnedKeys: {
                    next: false,
                },
            },
        });
        return futureResults;
    }
    return rules.map((rule, i) => {
        async function executeRPC(dataBox, toolbox) {
            return getOrFetchResults(rules, serializeParsersEngineDataBox(dataBox)).then(httpPayload => {
                const result = httpPayload[i];
                if (!result) {
                    throw new SpecifyError({
                        errorKey: specifyErrors.PARSERS_ENGINE_RPC_EXECUTION_FAILED.errorKey,
                        publicMessage: `RPC result not found for parsers "${rule.parsers
                            .map(p => p.name)
                            .join(', ')}"${rule.name ? ` on rule "${rule.name}"` : ''}.`,
                    });
                }
                const parserDefinition = getBuiltInParserDefinition(pickLastParserInParserRule(rule).name);
                return handleSerializedParsersEngineResult(httpPayload[i], toolbox, parserDefinition.outType);
            });
        }
        return executeRPC;
    });
}
