// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-css-text-style/definition.ts
import { z } from 'zod';
import { possibleTextStylesKeysSchema } from '../../shared/definition.js';
import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';

export const toCssTextStyleParserName = 'to-css-text-style';

export const toCssTextStyleParserOptionsSchema = z
  .object({
    tokenNameTemplate: z.string().optional(),
    exclude: possibleTextStylesKeysSchema.optional(),
    include: possibleTextStylesKeysSchema.optional(),
    genericFamily: z.string().optional(),
  })
  .strict();

export const toCssTextStyleParserDefinition = createBuiltInParserDefinition({
  name: toCssTextStyleParserName,
  kind: 'generation',
  hasOptionalOptions: true,
  optionsSchema: toCssTextStyleParserOptionsSchema,
  hasOptionalOutput: false,
  outputTypes: ['file'],
  inTypes: ['SDTF', 'SDTF Engine'],
  outType: 'SDTF Engine',
});

export type ToCssTextStyleParserDefinition = typeof toCssTextStyleParserDefinition;
export type ToCssTextStyleParserOptions = z.infer<ToCssTextStyleParserDefinition['optionsSchema']>;
