// @specifyapp/sdk/src/converters/css/tokenTypes/breakpoint.ts
import {
  SpecifyBreakpointValue,
  RawValueSignature,
  TokenState,
} from '@specifyapp/specify-design-token-format';
import { DEFAULT_UNRESOLVABLE_STRATEGY } from '../aliasStrategies/throwOnUnresolvable.js';
import {
  CssResolvableAliasStrategy,
  CssUnresolvableAliasStrategy,
} from '../aliasStrategies/CssAliasStrategy.js';

/**
 * Converts a `TokenState<'breakpoint'>` to a css value.
 * E.g: `12px`
 */
export function breakpointToCss<UnresolvableReturn = never>(
  aliasStrategy: CssResolvableAliasStrategy,
  unresolvableAliasStrategy: CssUnresolvableAliasStrategy<UnresolvableReturn> = DEFAULT_UNRESOLVABLE_STRATEGY,
) {
  return (tokenState: TokenState<'breakpoint'>) =>
    tokenState
      .getStatefulValueResult()
      .mapResolvableTopLevelAlias(aliasStrategy)
      .mapUnresolvableTopLevelAlias(alias => unresolvableAliasStrategy(tokenState, alias))
      .mapTopLevelValue(value =>
        value
          .mapRawValue(rawValue => rawBreakpointToCss(rawValue))
          .mapResolvableModeLevelAlias(aliasStrategy)
          .mapUnresolvableModeLevelAlias(alias => unresolvableAliasStrategy(tokenState, alias))
          .unwrap(),
      )
      .unwrap();
}

/**
 * Converts a `RawValueSignature<'dimension'>` to css.
 */
export function rawBreakpointToCss<UnresolvableReturn = undefined>(
  breakpoint: RawValueSignature<'dimension'>,
  unresolvableAliasStrategy: CssUnresolvableAliasStrategy<UnresolvableReturn> = DEFAULT_UNRESOLVABLE_STRATEGY,
) {
  return `${breakpoint.value
    .resolveDeepValue()
    .mapUnresolvableValueLevelAlias(alias =>
      unresolvableAliasStrategy(breakpoint.value.tokenState, alias),
    )
    .unwrap()}${breakpoint.unit
    .resolveDeepValue()
    .mapUnresolvableValueLevelAlias(alias =>
      unresolvableAliasStrategy(breakpoint.value.tokenState, alias),
    )
    .unwrap()}`;
}

/**
 * Converts a breakpoint to css.
 */
export function breakpointValueToCss(breakpoint: SpecifyBreakpointValue) {
  return `${breakpoint.value}${breakpoint.unit}`;
}
