// @specifyapp/sdk/src/converters/css/tokenTypes/radius.ts
import {
  SpecifyRadiusValue,
  RawValueSignature,
  TokenState,
} from '@specifyapp/specify-design-token-format';
import { DEFAULT_UNRESOLVABLE_STRATEGY } from '../aliasStrategies/throwOnUnresolvable.js';
import {
  CssResolvableAliasStrategy,
  CssUnresolvableAliasStrategy,
} from '../aliasStrategies/CssAliasStrategy.js';

/**
 * Converts a `TokenState<'radius'>` to a css value.
 * E.g: `12px`
 */
export function radiusToCss<UnresolvableReturn = never>(
  aliasStrategy: CssResolvableAliasStrategy,
  unresolvableAliasStrategy: CssUnresolvableAliasStrategy<UnresolvableReturn> = DEFAULT_UNRESOLVABLE_STRATEGY,
) {
  return (tokenState: TokenState<'radius'>) =>
    tokenState
      .getStatefulValueResult()
      .mapResolvableTopLevelAlias(aliasStrategy)
      .mapUnresolvableTopLevelAlias(alias => unresolvableAliasStrategy(tokenState, alias))
      .mapTopLevelValue(value =>
        value
          .mapRawValue(rawValue => rawRadiusToCss(rawValue, unresolvableAliasStrategy))
          .mapResolvableModeLevelAlias(aliasStrategy)
          .mapUnresolvableModeLevelAlias(alias => unresolvableAliasStrategy(tokenState, alias))
          .unwrap(),
      )
      .unwrap();
}

/**
 * Converts a `RawValueSignature<'radius'>` to css.
 */
export function rawRadiusToCss<UnresolvableReturn = never>(
  radius: RawValueSignature<'radius'>,
  unresolvableAliasStrategy: CssUnresolvableAliasStrategy<UnresolvableReturn> = DEFAULT_UNRESOLVABLE_STRATEGY,
) {
  return `${radius.value
    .resolveDeepValue()
    .mapUnresolvableValueLevelAlias(alias =>
      unresolvableAliasStrategy(radius.value.tokenState, alias),
    )
    .unwrap()}${radius.unit.resolveDeepValue().unwrapValue()}`;
}

/**
 * Converts a radius to css.
 */
export function radiusValueToCss(radius: SpecifyRadiusValue) {
  return `${radius.value}${radius.unit}`;
}
