// @specifyapp/specify-design-token-format/dist/engine/builder/createTokenState.js
import { SDTFError } from '../../errors/index.js';
import { TokenState } from '../state/TokenState.js';
export function createTokenState(treeState, analyzedToken) {
    if (analyzedToken.isFullyResolvable === undefined) {
        throw new SDTFError('SDTF_INTERNAL_DESIGN_ERROR', 'isFullyResolvable is undefined');
    }
    if (analyzedToken.modesResolvability === undefined) {
        throw new SDTFError('SDTF_INTERNAL_DESIGN_ERROR', 'modesResolvability is undefined');
    }
    return new TokenState(treeState, {
        path: analyzedToken.path,
        name: analyzedToken.name,
        $type: analyzedToken.$type,
        $description: analyzedToken.$description,
        $extensions: analyzedToken.$extensions,
        definition: analyzedToken.definition,
        isTopLevelAlias: analyzedToken.isTopLevelAlias,
        analyzedValuePrimitiveParts: analyzedToken.analyzedValuePrimitiveParts,
        isFullyResolvable: analyzedToken.isFullyResolvable,
        modesResolvabilityMap: new Map(Object.entries(analyzedToken.modesResolvability)),
    });
}
