// @specifyapp/sdk/dist/builtInParsers/parsers-generation/to-swift/toSwiftHandler.js
import { createSDTFEngine } from '@specifyapp/specify-design-token-format';
import { convertTokens } from './convertTokens.js';
import { convertedToSwift } from './convertedToSwift.js';
import { DEFAULT_TEMPLATE, validateTemplate } from './template.js';
import { SpecifyError, specifyErrors } from '../../../errors/index.js';
export const toSwiftHandler = async (previousDataBox, toolbox, parserOptions, outputOptions, _context) => {
    let sdtfEngine;
    switch (previousDataBox.type) {
        case 'SDTF': {
            sdtfEngine = createSDTFEngine(previousDataBox.graph, previousDataBox.metadata);
            break;
        }
        case 'SDTF Engine': {
            sdtfEngine = previousDataBox.engine;
            break;
        }
        default: {
            throw new SpecifyError({
                errorKey: specifyErrors.PARSERS_ENGINE_INVALID_PARSER_INPUT.errorKey,
                publicMessage: `${previousDataBox.type} is not a valid input for the to-swift parser.`,
            });
        }
    }
    if (outputOptions?.type !== 'file')
        throw new SpecifyError({
            errorKey: specifyErrors.PARSERS_ENGINE_INVALID_OUTPUT_TYPE.errorKey,
            publicMessage: `The output type ${outputOptions?.type} is not supported by the to-swift parser.`,
        });
    // Will throw if template is not valid
    if (parserOptions?.tokenNameTemplate)
        validateTemplate(parserOptions.tokenNameTemplate);
    toolbox.populateOutput({
        type: 'files',
        files: [
            {
                path: outputOptions.filePath,
                content: {
                    type: 'text',
                    text: convertedToSwift(convertTokens(sdtfEngine.query.getAllTokenStates(), parserOptions?.tokenNameTemplate ?? DEFAULT_TEMPLATE), parserOptions?.scopeName ?? 'DesignTokens'),
                },
            },
        ],
    });
    return {
        type: 'SDTF Engine',
        engine: sdtfEngine,
    };
};
