// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-css-font-import/definition.ts
import { z } from 'zod';

import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';

export const toCssFontImportParserName = 'to-css-font-import';

export const toCssFontImportParserOptionsSchema = z
  .object({
    formats: z
      .array(
        z.union([
          z.literal('woff'),
          z.literal('woff2'),
          z.literal('ttf'),
          z.literal('otf'),
          z.literal('eot'),
        ]),
      )
      .optional(),
    fontsPath: z.string().optional(),
    includeFontWeight: z.boolean().optional(),
    genericFamily: z.string().optional(),
    fontDisplay: z
      .union([
        z.literal('auto'),
        z.literal('block'),
        z.literal('swap'),
        z.literal('fallback'),
        z.literal('optional'),
      ])
      .optional(),
  })
  .strict();

export const toCssFontImportParserDefinition = createBuiltInParserDefinition({
  name: toCssFontImportParserName,
  kind: 'generation',
  hasOptionalOptions: true,
  optionsSchema: toCssFontImportParserOptionsSchema,
  hasOptionalOutput: false,
  outputTypes: ['file'],
  inTypes: ['SDTF', 'SDTF Engine'],
  outType: 'SDTF Engine',
});
export type ToCssFontImportParserDefinition = typeof toCssFontImportParserDefinition;
export type ToCssFontImportParserOptions = z.infer<
  ToCssFontImportParserDefinition['optionsSchema']
>;
