// @specifyapp/sdk/src/builtInParsers/parsers-utility/convert-dimension/tokenTypes/border.ts
import {
  RawValueSignature,
  SpecifyDimensionValue,
  TokenState,
} from '@specifyapp/specify-design-token-format';
import type { options } from '../updateToken.js';
import { getDimensionUpdateValueOrUpdateAlias } from './dimension.js';
import { getRadiiUpdateValueOrUpdateAlias } from './radii.js';
import { ParserToolbox } from '../../../../parsersEngine/ParserToolbox.js';
import { KeysPerToken } from '../definition.js';

export function getBorderUpdateValue(
  border: RawValueSignature<'border'>,
  options: options,
  token: TokenState<'border'>,
  mode: string,
  toolbox: ParserToolbox,
  allowedKeys: KeysPerToken['border'],
) {
  const initialValueToUpdate = token.getJSONValue({ resolveAliases: false });

  if ('$alias' in initialValueToUpdate)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  const valueToUpdate = initialValueToUpdate[mode];

  if ('$alias' in valueToUpdate)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  if (!allowedKeys || allowedKeys.includes('width')) {
    const width = getDimensionUpdateValueOrUpdateAlias(border.width, options, toolbox);

    if (!!width) valueToUpdate.width = width;
  }

  if (!allowedKeys || allowedKeys.includes('rectangleCornerRadii')) {
    const rectangleCornerRadii = getRadiiUpdateValueOrUpdateAlias(
      border.rectangleCornerRadii,
      options,
      toolbox,
    );

    if (!!rectangleCornerRadii)
      valueToUpdate.rectangleCornerRadii = rectangleCornerRadii as [
        SpecifyDimensionValue,
        SpecifyDimensionValue,
        SpecifyDimensionValue,
        SpecifyDimensionValue,
      ];
  }

  return valueToUpdate;
}
