// @specifyapp/sdk/src/builtInParsers/parsers-utility/convert-dimension/tokenTypes/dimension.ts
import {
  InnerValue,
  RawValueSignature,
  ResolvableValueLevelAlias,
  UnresolvableModeLevelAlias,
  UnresolvableValueLevelAlias,
  ResolvedDeepStatefulValueForMode,
} from '@specifyapp/specify-design-token-format';
import type { options } from '../updateToken.js';
import { convertDimension } from '../conversion.js';
import { specifyErrors } from '../../../../errors/index.js';
import { ParserToolbox } from '../../../../parsersEngine/ParserToolbox.js';

export function getDimensionUpdateValueOrUpdateAlias(
  dimension: InnerValue<
    ResolvableValueLevelAlias | UnresolvableValueLevelAlias | RawValueSignature<'dimension'> | null
  >,
  options: options,
  toolbox: ParserToolbox,
) {
  return dimension
    .mapPrimitiveValue(v => (!!v ? getDimensionUpdateValue(v, options, toolbox) : undefined))
    .mapUnresolvableValueLevelAlias(_ => undefined)
    .mapResolvableValueLevelAlias(alias => {
      const { tokenState, targetMode, value } = alias.tokenState.resolveDeepStatefulValueForMode(
        alias.targetMode,
      ) as ResolvedDeepStatefulValueForMode<'dimension'>;

      if (value instanceof UnresolvableModeLevelAlias)
        throw new Error('Unreachable as token resolvability is checked before');

      tokenState.updateModeValue(targetMode, getDimensionUpdateValue(value, options, toolbox));

      return undefined;
    })
    .unwrap();
}

export function getDimensionUpdateValue(
  dimension: RawValueSignature<'dimension'>,
  options: options,
  toolbox: ParserToolbox,
) {
  const unit = dimension.unit.resolveDeepValue().unwrapValue();
  const value = dimension.value.resolveDeepValue().unwrapValue();

  if (options.includeFormats && !options.includeFormats.includes(unit)) return { unit, value };

  if (options.excludeFormats && options.excludeFormats.includes(unit)) return { unit, value };

  const convertedDimension = convertDimension(
    { unit, value },
    options.targetUnit,
    options.baseValue,
  );

  if (!convertedDimension) {
    toolbox.populateMessage({
      type: 'warning',
      content: `Tried to convert '${unit}' to '${options.targetUnit}', the conversion is not compatible, or not compatible yet. Please contact us if you need that conversion`,
      errorKey: specifyErrors.PARSERS_ENGINE_PARSER_EXECUTION_FAILED.errorKey,
    });
    return { unit, value };
  }

  return convertedDimension;
}
