// @specifyapp/sdk/src/builtInParsers/parsers-utility/convert-dimension/tokenTypes/shadows.ts
import { RawValueSignature, TokenState } from '@specifyapp/specify-design-token-format';
import type { options } from '../updateToken.js';
import {
  getShadowUpdateValueFromValueToUpdate,
  getShadowUpdateValueOrUpdateAliasFromValueToUpdate,
} from './shadow.js';
import { ParserToolbox } from '../../../../parsersEngine/ParserToolbox.js';
import { KeysPerToken } from '../definition.js';

export function getShadowsUpdateValue(
  shadows: RawValueSignature<'shadows'>,
  options: options,
  token: TokenState<'shadows'>,
  mode: string,
  toolbox: ParserToolbox,
  allowedKeys: KeysPerToken['shadow'],
) {
  const rawValue = token.getJSONValue({ resolveAliases: false });

  if ('$alias' in rawValue)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  const valuesToUpdate = rawValue[mode];

  if ('$alias' in valuesToUpdate)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  return shadows.map((innerValue, i) =>
    innerValue
      .mapPrimitiveValue(shadow =>
        getShadowUpdateValueFromValueToUpdate(
          shadow,
          options,
          valuesToUpdate[i],
          toolbox,
          allowedKeys,
        ),
      )
      .mapUnresolvableValueLevelAlias(_ => {
        throw new Error('Unreachable as token resolvability is checked before');
      })
      .mapResolvableValueLevelAlias(alias => {
        const { tokenState, targetMode } = alias.tokenState.resolveDeepStatefulValueForMode(
          alias.targetMode,
        );

        const topLevel = (tokenState as TokenState<'shadow'>).getJSONValue({
          resolveAliases: false,
        });

        if ('$alias' in topLevel)
          throw new Error('Unreachable as we dug to the value from the mode level');

        const valueToUpdate = topLevel[targetMode];

        if ('$alias' in valuesToUpdate) throw new Error('Unreachable as we dug to the value');
        // We update the alias, and we return the format expected by the SDTF to update the token
        getShadowUpdateValueOrUpdateAliasFromValueToUpdate(
          innerValue,
          options,
          valueToUpdate,
          toolbox,
          allowedKeys,
        );

        return { $alias: alias.tokenState.path.toString(), $mode: alias.targetMode };
      })
      .unwrap(),
  );
}
