// @specifyapp/sdk/src/builtInParsers/parsers-utility/convert-dimension/tokenTypes/textStyle.ts
import { RawValueSignature, TokenState } from '@specifyapp/specify-design-token-format';
import type { options } from '../updateToken.js';
import { getDimensionUpdateValueOrUpdateAlias } from './dimension.js';
import { ParserToolbox } from '../../../../parsersEngine/ParserToolbox.js';
import { KeysPerToken } from '../definition.js';

export function getTextStyleUpdateValue(
  textStyle: RawValueSignature<'textStyle'>,
  options: options,
  token: TokenState<'textStyle'>,
  mode: string,
  toolbox: ParserToolbox,
  allowedKeys: KeysPerToken['textStyle'],
) {
  const initialValueToUpdate = token.getJSONValue({ resolveAliases: false });

  if ('$alias' in initialValueToUpdate)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  const valueToUpdate = initialValueToUpdate[mode];

  if ('$alias' in valueToUpdate)
    throw new Error("Unreachable as it's checked before by the statefulResult api");

  if (!allowedKeys || allowedKeys.includes('fontSize')) {
    const fontSize = getDimensionUpdateValueOrUpdateAlias(textStyle.fontSize, options, toolbox);
    if (!!fontSize) valueToUpdate.fontSize = fontSize;
  }

  if (!allowedKeys || allowedKeys.includes('lineHeight')) {
    const lineHeight = getDimensionUpdateValueOrUpdateAlias(textStyle.lineHeight, options, toolbox);
    if (!!lineHeight) valueToUpdate.lineHeight = lineHeight;
  }

  if (!allowedKeys || allowedKeys.includes('textIndent')) {
    const textIndent = getDimensionUpdateValueOrUpdateAlias(textStyle.textIndent, options, toolbox);
    if (!!textIndent) valueToUpdate.textIndent = textIndent;
  }

  if (!allowedKeys || allowedKeys.includes('letterSpacing')) {
    const letterSpacing = getDimensionUpdateValueOrUpdateAlias(
      textStyle.letterSpacing,
      options,
      toolbox,
    );
    if (!!letterSpacing) valueToUpdate.letterSpacing = letterSpacing;
  }

  if (!allowedKeys || allowedKeys.includes('paragraphSpacing')) {
    const paragraphSpacing = getDimensionUpdateValueOrUpdateAlias(
      textStyle.paragraphSpacing,
      options,
      toolbox,
    );
    if (!!paragraphSpacing) valueToUpdate.paragraphSpacing = paragraphSpacing;
  }

  return valueToUpdate;
}
