// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-react-native/tokenTypes/textStyle.ts
import { SpecifyTextStyleValue } from '@specifyapp/specify-design-token-format';
import { renderColor } from './color.js';
import { renderDimension } from './dimension.js';
import { renderFont } from './font.js';

export function renderTextStyle(textStyle: SpecifyTextStyleValue) {
  const variables = renderFont(textStyle.font) as ReturnType<typeof renderFont> & {
    fontSize: string | number;
    color?: string;
    textIndent?: string | number;
    lineHeight?: string | number;
    letterSpacing?: string | number;
    textAlignHorizontal?: string;
    textAlignVertical?: string;
    textTransform?: string;
    textDecoration?: string;
  };

  variables.fontSize = renderDimension(textStyle.fontSize);

  if (!!textStyle.color) variables.color = renderColor(textStyle.color);

  if (!!textStyle.textIndent) variables.textIndent = renderDimension(textStyle.textIndent);

  if (!!textStyle.lineHeight) variables.lineHeight = renderDimension(textStyle.lineHeight);

  if (!!textStyle.letterSpacing) variables.letterSpacing = renderDimension(textStyle.letterSpacing);

  if (!!textStyle.textAlignHorizontal)
    variables.textAlignHorizontal = textStyle.textAlignHorizontal;

  if (!!textStyle.textAlignVertical) variables.textAlignVertical = textStyle.textAlignVertical;

  if (!!textStyle.textTransform) variables.textTransform = textStyle.textTransform;

  if (!!textStyle.textDecoration) variables.textDecoration = textStyle.textDecoration;

  return variables;
}
