// @specifyapp/specify-design-token-format/dist/engine/query/internals/nodePropertiesMatching.js
import { z } from 'zod';
export const nodePropertiesMatchingObjectSchema = z
    .object({
    name: z.string().optional(),
    description: z.string().optional(),
})
    .refine(value => {
    return value.name !== undefined || value.description !== undefined;
}, {
    message: 'Where group | token | collection object must include either `name` or `description`',
});
export function matchIsNodePropertiesMatchingObject(match) {
    if (!Array.isArray(match) && typeof match === 'object' && match !== null) {
        return Object.keys(match).every(key => ['name', 'description'].includes(key));
    }
    return false;
}
export const nodePropertiesMatchingSchema = z.union([
    z.string(),
    nodePropertiesMatchingObjectSchema,
]);
export function makeMaybeNodePropertiesRegexes(nodeDescriptor) {
    let maybeNameRegex;
    let maybeDescriptionRegex;
    if (matchIsNodePropertiesMatchingObject(nodeDescriptor)) {
        if (nodeDescriptor.name) {
            maybeNameRegex = new RegExp(nodeDescriptor.name);
        }
        if (nodeDescriptor.description) {
            maybeDescriptionRegex = new RegExp(nodeDescriptor.description);
        }
    }
    else {
        maybeNameRegex = new RegExp(nodeDescriptor);
    }
    return { maybeNameRegex, maybeDescriptionRegex };
}
