// @specifyapp/specify-design-token-format/dist/engine/state/TreeNodeSet.js
import { makeGetChildrenOfFilter } from '../utils/makeGetChildrenOfFilter.js';
import { makeGetParentOfFilter } from '../utils/makeGetParentOfFilter.js';
export class TreeNodeSet {
    #nodes = [];
    static makeStringPath(path) {
        return path.toString();
    }
    constructor(nodes) {
        if (nodes) {
            this.#nodes.push(...nodes);
        }
    }
    get all() {
        return this.#nodes;
    }
    get size() {
        return this.#nodes.length;
    }
    *[Symbol.iterator]() {
        const data = this.#nodes;
        for (let i = 0; i < data.length; i++) {
            yield data[i];
        }
    }
    add(...nodes) {
        this.#nodes.push(...nodes);
    }
    delete(path) {
        const targetStringPath = TreeNodeSet.makeStringPath(path);
        const index = this.#nodes.findIndex(node => node.path.toString() === targetStringPath);
        if (index === -1) {
            return false;
        }
        this.#nodes.splice(index, 1);
        return true;
    }
    clear() {
        this.#nodes.length = 0;
    }
    has(path) {
        return this.#nodes.some(node => node.path.toString() === path.toString());
    }
    getOne(path) {
        const targetStringPath = TreeNodeSet.makeStringPath(path);
        return this.#nodes.find(node => node.path.toString() === targetStringPath);
    }
    getChildrenOf(path, depth = Infinity) {
        return this.#nodes.filter(makeGetChildrenOfFilter(path, depth));
    }
    getParentsOf(path, depth = Infinity) {
        return this.#nodes.filter(makeGetParentOfFilter(path, depth)).reverse().splice(0, depth);
    }
}
