// @specifyapp/specify-design-token-format/dist/engine/state/ViewState.js
import { z } from 'zod';
import { makeRunQuery, parseQuery } from '../query/index.js';
import { TreeNodesState } from './TreeNodesState.js';
export const serializedViewSchema = z.object({
    name: z.string(),
    query: z
        .custom(v => v)
        .superRefine((value, ctx) => {
        try {
            parseQuery(value);
        }
        catch (error) {
            if (error instanceof z.ZodError) {
                // TODO @Nico: improve error mapping
                error.issues.forEach(issue => {
                    ctx.addIssue({
                        code: z.ZodIssueCode.custom,
                        message: issue.message,
                    });
                });
            }
            else {
                const message = typeof error === 'object' &&
                    error !== null &&
                    'message' in error &&
                    typeof error.message === 'string'
                    ? error.message
                    : 'Unknown SDTF query parsing error';
                ctx.addIssue({
                    code: z.ZodIssueCode.custom,
                    message: message,
                });
            }
        }
    }),
});
export class ViewState {
    #name;
    #query;
    #nodes;
    constructor(name, query, nodes = new TreeNodesState()) {
        this.#name = name;
        this.#query = query;
        this.#nodes = nodes;
    }
    get name() {
        return this.#name;
    }
    get query() {
        return this.#query;
    }
    get nodes() {
        return this.#nodes;
    }
    updateQuery(query, treeState) {
        this.#query = query;
        this.refresh(treeState);
    }
    refresh(treeState) {
        this.#nodes.clear();
        const results = makeRunQuery(treeState)(this.#query);
        for (const node of results) {
            if (node.isToken) {
                this.#nodes.tokens.add(node);
            }
            else if (node.isGroup) {
                this.#nodes.groups.add(node);
            }
            else if (node.isCollection) {
                this.#nodes.collections.add(node);
            }
        }
    }
    serialize() {
        return {
            name: this.#name,
            query: this.#query,
        };
    }
}
