// @specifyapp/sdk/dist/converters/scss/aliasStrategies/resolvableResolve.js
import { ResolvableTopLevelAlias } from '@specifyapp/specify-design-token-format';
import { transitionValueToCss } from '../../css/tokenTypes/transition.js';
import { dimensionValueToCss } from '../../css/tokenTypes/dimension.js';
import { breakpointValueToCss } from '../../css/tokenTypes/breakpoint.js';
import { colorValueToCss } from '../../css/tokenTypes/color.js';
import { cubicBezierValueToCss } from '../../css/tokenTypes/cubicBezier.js';
import { durationValueToCss } from '../../css/tokenTypes/duration.js';
import { zIndexValueToCss } from '../../css/tokenTypes/zIndex.js';
import { borderValueToCss } from '../../css/tokenTypes/border.js';
import { fontWeightValueToCss } from '../../css/tokenTypes/fontWeight.js';
import { gradientValueToCss } from '../../css/tokenTypes/gradient.js';
import { gradientsValueToCss } from '../../css/tokenTypes/gradients.js';
import { opacityValueToCss } from '../../css/tokenTypes/opacity.js';
import { radiusValueToCss } from '../../css/tokenTypes/radius.js';
import { radiiValueToCss } from '../../css/tokenTypes/radii.js';
import { shadowValueToCss } from '../../css/tokenTypes/shadow.js';
import { shadowsValueToCss } from '../../css/tokenTypes/shadows.js';
import { spacingValueToCss } from '../../css/tokenTypes/spacing.js';
import { spacingsValueToCss } from '../../css/tokenTypes/spacings.js';
import { stepsTimingFunctionValueToCss } from '../../css/tokenTypes/stepsTimingFunction.js';
import { textStyleValueToCss } from '../../css/tokenTypes/textStyle.js';
import { fontValueToCss } from '../../css/tokenTypes/font.js';
function getJsonValue(tokenState, mode) {
    return tokenState.matchJSONValueByType({
        dimension: dimensionValueToCss,
        breakpoint: breakpointValueToCss,
        color: colorValueToCss,
        cubicBezier: cubicBezierValueToCss,
        duration: durationValueToCss,
        zIndex: zIndexValueToCss,
        border: borderValueToCss,
        font: fontValueToCss,
        fontWeight: fontWeightValueToCss,
        gradient: gradientValueToCss,
        gradients: gradientsValueToCss,
        opacity: opacityValueToCss,
        radius: radiusValueToCss,
        radii: radiiValueToCss,
        shadow: shadowValueToCss,
        shadows: shadowsValueToCss,
        spacing: spacingValueToCss,
        spacings: spacingsValueToCss,
        stepsTimingFunction: stepsTimingFunctionValueToCss,
        textStyle: textStyleValueToCss,
        transition: transitionValueToCss,
    }, _ => {
        throw new Error(`Token of type: ${tokenState.type} can't be converted to css`);
    })[mode];
}
/**
 * Will create a resolve alias strategy. By applying this strategy when converting a token to css,
 * rather than converting the alias to a css variable, we will resolve it and convert it to the css value, e.g: `12px`
 */
export function createResolveAliasStrategy() {
    return value => {
        if (value instanceof ResolvableTopLevelAlias) {
            return value.tokenState.modes.reduce((acc, curr) => {
                const converted = getJsonValue(value.tokenState, curr);
                if (!converted)
                    return acc;
                acc[curr] = converted;
                return acc;
            }, {});
        }
        else {
            return getJsonValue(value.tokenState, value.targetMode);
        }
    };
}
