// @specifyapp/specify-design-token-format/dist/engine/query/definitions/collectionWhere.js
import { z } from 'zod';
import { childrenAndParentsUpToDepthOperatorSchema } from '../internals/childrenAndParentsUpToDepthOperator.js';
import { nodePropertiesMatchingSchema } from '../internals/nodePropertiesMatching.js';
import { withModesSchema } from '../internals/withModes.js';
import { nodeWithAtDepthSchema } from '../internals/nodeAtDepthProperty.js';
export const tokenChildrenOfCollectionUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('group')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const groupChildrenOfCollectionUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('token')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const groupParentsOfCollectionUpToDepthOperatorSchema = childrenAndParentsUpToDepthOperatorSchema;
export const collectionSelectSchema = z.object({
    collection: z.boolean().optional(),
    children: z
        .union([
        z.literal(true),
        childrenAndParentsUpToDepthOperatorSchema,
        z
            .object({
            tokens: z
                .union([z.literal(true), tokenChildrenOfCollectionUpToDepthOperatorSchema])
                .optional(),
            groups: z
                .union([z.literal(true), groupChildrenOfCollectionUpToDepthOperatorSchema])
                .optional(),
        })
            .strict(),
    ])
        .optional(),
    parents: z
        .union([
        z.literal(true),
        childrenAndParentsUpToDepthOperatorSchema,
        z
            .object({
            groups: z
                .union([z.literal(true), groupParentsOfCollectionUpToDepthOperatorSchema])
                .optional(),
        })
            .strict(),
    ])
        .optional(),
});
export const collectionSelectPropertySchema = z.union([z.literal(true), collectionSelectSchema]);
const collectionWithSelectSchema = z.object({
    select: collectionSelectPropertySchema,
});
const collectionNestedInPropertySchema = z.object({
    group: z.boolean().optional(),
});
const collectionWithNestedInSchema = z.object({
    nestedIn: collectionNestedInPropertySchema.optional(),
});
const collectionWithSelfSelectorSchema = z.object({
    collection: nodePropertiesMatchingSchema,
});
export const collectionWhereBaseSchema = collectionWithSelfSelectorSchema
    .merge(nodeWithAtDepthSchema)
    .merge(collectionWithNestedInSchema)
    .merge(withModesSchema);
export function matchIsCollectionWhere(where) {
    return 'collection' in where;
}
export function matchIsCollectionWhereWithSelect(where) {
    return 'select' in where && matchIsCollectionWhere(where);
}
export function matchIsCollectionWhereWithAndWhere(where) {
    return 'andWhere' in where && matchIsCollectionWhere(where);
}
