// @specifyapp/specify-design-token-format/dist/engine/query/definitions/groupWhere.js
import { z } from 'zod';
import { childrenAndParentsUpToDepthOperatorSchema } from '../internals/childrenAndParentsUpToDepthOperator.js';
import { nodePropertiesMatchingSchema } from '../internals/nodePropertiesMatching.js';
import { nodeWithAtDepthSchema } from '../internals/nodeAtDepthProperty.js';
export const tokenChildrenOfGroupUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('group'), z.literal('collection')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const groupChildrenOfGroupUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('collection'), z.literal('token')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const collectionChildrenOfGroupUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('group'), z.literal('token')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const groupParentsOfGroupUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('collection')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const collectionParentsOfGroupUpToDepthOperatorSchema = z.union([
    z
        .object({
        upToDepth: z.union([z.number().min(0), z.literal('group')]),
        equalToDepth: z.undefined(),
    })
        .strict(),
    z
        .object({
        upToDepth: z.undefined(),
        equalToDepth: z.number().min(0),
    })
        .strict(),
]);
export const groupSelectSchema = z
    .object({
    group: z.boolean().optional(),
    children: z
        .union([
        z.literal(true),
        childrenAndParentsUpToDepthOperatorSchema,
        z
            .object({
            tokens: z
                .union([z.literal(true), tokenChildrenOfGroupUpToDepthOperatorSchema])
                .optional(),
            groups: z
                .union([z.literal(true), groupChildrenOfGroupUpToDepthOperatorSchema])
                .optional(),
            collections: z
                .union([z.literal(true), collectionChildrenOfGroupUpToDepthOperatorSchema])
                .optional(),
        })
            .strict(),
    ])
        .optional(),
    parents: z
        .union([
        z.literal(true),
        childrenAndParentsUpToDepthOperatorSchema,
        z
            .object({
            groups: z
                .union([z.literal(true), groupParentsOfGroupUpToDepthOperatorSchema])
                .optional(),
            collections: z
                .union([z.literal(true), collectionParentsOfGroupUpToDepthOperatorSchema])
                .optional(),
        })
            .strict(),
    ])
        .optional(),
})
    .strict();
export const groupSelectPropertySchema = z.union([z.literal(true), groupSelectSchema]);
export const groupWithSelectSchema = z.object({
    select: groupSelectPropertySchema,
});
const groupNestedInPropertySchema = z.object({
    collection: z.boolean().optional(),
    group: z.boolean().optional(),
});
const groupWithNestedInSchema = z.object({
    nestedIn: groupNestedInPropertySchema.optional(),
});
const groupWithSelfSelectorSchema = z.object({
    group: nodePropertiesMatchingSchema,
});
export const groupWhereBaseSchema = groupWithSelfSelectorSchema
    .merge(nodeWithAtDepthSchema)
    .merge(groupWithNestedInSchema);
export function matchIsGroupWhere(where) {
    return 'group' in where;
}
export function matchIsGroupWhereWithSelect(where) {
    return 'select' in where && matchIsGroupWhere(where);
}
export function matchIsGroupWhereWithAndWhere(where) {
    return 'andWhere' in where && matchIsGroupWhere(where);
}
