// @specifyapp/sdk/dist/client/internals/Authentication.js
import { SpecifyError, specifyErrors } from '../../errors/index.js';
export class Authentication {
    #currentUser;
    #personalAccessToken;
    get isAuthenticated() {
        return this.#currentUser !== undefined;
    }
    get credentials() {
        if (!this.#currentUser || !this.#personalAccessToken) {
            throw new SpecifyError({
                publicMessage: 'No current user authenticated. Please authenticate first',
                errorKey: specifyErrors.AUTHENTICATION_UNEXPECTED_ERROR.errorKey,
            });
        }
        return {
            personalAccessToken: this.#personalAccessToken,
            currentUser: this.#currentUser,
            currentOrganization: this.#currentUser.organizations[0],
        };
    }
    get personalAccessToken() {
        return this.credentials.personalAccessToken;
    }
    set(personalAccessToken, currentUser) {
        this.#currentUser = currentUser;
        this.#personalAccessToken = personalAccessToken;
    }
    reset() {
        this.#currentUser = undefined;
        this.#personalAccessToken = undefined;
    }
}
