// @specifyapp/sdk/dist/builtInParsers/parsers-generation/to-javascript/tokenTypes/color.js
//@ts-ignore
import tinycolor from 'tinycolor2';
export function renderColor(color) {
    const alpha = color.alpha;
    switch (color.model) {
        case 'hex': {
            return renderHex(color.hex, alpha);
        }
        case 'rgb': {
            const tc = tinycolor({
                r: color.red,
                g: color.green,
                b: color.blue,
            });
            tc.setAlpha(alpha);
            return tc.toRgbString();
        }
        case 'hsl': {
            const tc = tinycolor({
                h: color.hue,
                s: color.saturation,
                l: color.lightness,
            });
            tc.setAlpha(alpha);
            return tc.toHslString();
        }
        case 'hsb': {
            const tc = tinycolor({
                h: color.hue,
                s: color.saturation,
                v: color.brightness,
            });
            tc.setAlpha(alpha);
            return tc.toHsvString();
        }
        case 'lch': {
            return `lch(${color.lightness}% ${color.chroma} ${color.hue}${alpha < 1 ? ' / ' + alpha : ''})`;
        }
        case 'lab': {
            return `lab(${color.lightness}% ${color.aAxis} ${color.bAxis}${alpha < 1 ? ' / ' + alpha : ''})`;
        }
    }
}
function renderHex(hex, alpha) {
    const tc = tinycolor(hex);
    if (alpha === 1) {
        return tc.toHexString();
    }
    tc.setAlpha(alpha);
    return tc.toHex8String();
}
