// @specifyapp/sdk/dist/builtInParsers/parsers-generation/to-javascript/tokenTypes/textStyle.js
import { renderColor } from './color.js';
import { renderDimension } from './dimension.js';
import { renderFont } from './font.js';
export function renderTextStyle(textStyle) {
    const variables = renderFont(textStyle.font);
    variables['fontSize'] = renderDimension(textStyle.fontSize);
    if (!!textStyle.color)
        variables['color'] = renderColor(textStyle.color);
    if (!!textStyle.textIndent)
        variables['textIndent'] = renderDimension(textStyle.textIndent);
    if (!!textStyle.lineHeight)
        variables['lineHeight'] = renderDimension(textStyle.lineHeight);
    if (!!textStyle.letterSpacing)
        variables['letterSpacing'] = renderDimension(textStyle.letterSpacing);
    if (!!textStyle.textAlignHorizontal)
        variables['textAlign'] = textStyle.textAlignHorizontal;
    if (!!textStyle.textTransform)
        variables['textTransform'] = textStyle.textTransform;
    if (!!textStyle.textDecoration)
        variables['textDecoration'] = textStyle.textDecoration;
    return variables;
}
