// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-scss-map/tokenTypes/dimension.ts
import { TokenState } from '@specifyapp/specify-design-token-format';
import { dimensionValueToCss } from '../../../../converters/css/tokenTypes/dimension.js';
import { dataOfToken, makeRenderer } from '../../../shared/to-scss-mixin-text-style/template.js';
import { makeFunction } from '../makeFunction.js';

export function dimensionsToScssMap(
  type: string,
  dimensions: Array<TokenState<'dimension'> | TokenState<'spacing'> | TokenState<'breakpoint'>>,
  template: string,
) {
  const renderVariable = makeRenderer(template);

  let map = dimensions.reduce((acc, dimension) => {
    const toAppend = dimension.modes.reduce((acc, mode) => {
      const jsonValue = dimension.getJSONValue({
        resolveAliases: true,
        allowUnresolvable: false,
        targetMode: mode,
      });

      const converted = dimensionValueToCss(jsonValue);

      const varName = renderVariable(dataOfToken(dimension, mode));

      return acc + `  ${varName}: ${converted},\n`;
    }, '');

    return acc + toAppend;
  }, `$${type}: (\n`);

  map += ');\n';

  return map + makeFunction(type);
}
