// @specifyapp/sdk/dist/builtInParsers/parsers-generation/to-css-custom-properties/definition.js
import { z } from 'zod';
import { createBuiltInParserDefinition } from '../../internals/createBuiltInParserDefinition.js';
import { withSDTFViewSchema } from '../../shared/definition.js';
export const toCssCustomPropertiesParserName = 'to-css-custom-properties';
export const toCssCustomPropertiesParserOptionsSchema = z
    .object({
    selectorTemplate: z.string().optional(),
    tokenNameTemplate: z.string().optional(),
    tokenNotInCollectionNameTemplate: z.string().optional(),
    includeCoreTokensInScopes: z.boolean().optional(),
    allowUnresolvable: z.boolean().optional(),
    withSDTFView: withSDTFViewSchema,
})
    .strict();
export const toCssCustomPropertiesParserDefinition = createBuiltInParserDefinition({
    name: toCssCustomPropertiesParserName,
    kind: 'generation',
    hasOptionalOptions: true,
    optionsSchema: toCssCustomPropertiesParserOptionsSchema,
    hasOptionalOutput: false,
    outputTypes: ['file'],
    inTypes: ['SDTF Engine', 'SDTF'],
    outType: 'SDTF Engine',
});
