// @specifyapp/sdk/dist/builtInParsers/utils/template.js
import { ResolvableTopLevelAlias, } from '@specifyapp/specify-design-token-format';
import handlebars from 'handlebars';
handlebars.registerHelper('join', function (array, separator) {
    return array.join(separator);
});
export const dataOfToken = (token, mode) => {
    const groups = token.path.length <= 1 ? [] : token.path.toArray().slice(0, token.path.length - 1);
    const collectionState = token.getCollection();
    const collection = collectionState?.name;
    let groupsBeforeCollection = groups.length > 0 && collectionState ? groups.slice(0, collectionState.path.length - 1) : [];
    let groupsAfterCollection = groups.length > 0 && collectionState ? groups.slice(collectionState.path.length) : [];
    return {
        groups,
        path: groups,
        groupList: groups,
        groupsBeforeCollection,
        groupsAfterCollection,
        collection,
        mode,
        token: token.name,
    };
};
export const dataOfResolvedStatefulAlias = (resolved, mode) => dataOfToken(resolved.tokenState, resolved instanceof ResolvableTopLevelAlias ? mode : resolved.targetMode);
export function dataOfUnresolvedAlias(alias) {
    const token = alias.targetPath.at(alias.targetPath.length - 1);
    const path = alias.targetPath.length > 1
        ? alias.targetPath.toArray().slice(0, alias.targetPath.length - 1)
        : [];
    return {
        path,
        groups: path,
        groupList: path,
        collection: undefined,
        groupsBeforeCollection: [],
        groupsAfterCollection: [],
        token,
        mode: alias.targetMode,
    };
}
export function renderTemplate(template, data) {
    const render = handlebars.compile(template);
    return render(data);
}
export function sanitizeData(data, invalidChars, replacer) {
    const groupList = data.groupList.map(group => group.replaceAll(invalidChars, replacer));
    const groups = data.groups.length > 0
        ? data.groups.join(replacer).replaceAll(invalidChars, replacer)
        : undefined;
    const groupsBeforeCollectionList = data.groupsBeforeCollection.map(group => group.replaceAll(invalidChars, replacer));
    const groupsAfterCollectionList = data.groupsAfterCollection.map(group => group.replaceAll(invalidChars, replacer));
    return {
        token: data.token.replaceAll(invalidChars, replacer),
        groups,
        groupList,
        path: groups,
        collection: data.collection?.replaceAll(invalidChars, replacer),
        groupsBeforeCollectionList,
        groupsAfterCollectionList,
        groupsBeforeCollection: groupsBeforeCollectionList.join(replacer),
        groupsAfterCollection: groupsAfterCollectionList.join(replacer),
        mode: data.mode?.replaceAll(invalidChars, replacer),
    };
}
