// @specifyapp/specify-design-token-format/dist/definitions/internals/createDesignTokenDefinition.js
import { matchTokenTypeAgainstMapping, validateTokenTypesMapping, } from './tokenTypesMapping.js';
import { makeSpecifyDesignTokenSchema } from './designTokenSignature.js';
export function createDesignTokenDefinition(params) {
    const computedTypesMapping = validateTokenTypesMapping(!params.tokenTypesMapping ? { _tokenType: params.type } : params.tokenTypesMapping);
    const aliasableTokenZodSchema = makeSpecifyDesignTokenSchema(params.type, params.aliasableValueZodSchema, true);
    const resolvedTokenZodSchema = makeSpecifyDesignTokenSchema(params.type, params.resolvedValueZodSchema, false);
    return {
        type: params.type,
        tokenTypesMapping: computedTypesMapping,
        aliasableValueZodSchema: params.aliasableValueZodSchema,
        resolvedValueZodSchema: params.resolvedValueZodSchema,
        aliasableTokenZodSchema,
        resolvedTokenZodSchema,
        matchTokenTypeAgainstMapping: (type, path, getDiscriminatorValue = () => undefined) => {
            return matchTokenTypeAgainstMapping(type, computedTypesMapping, path, getDiscriminatorValue);
        },
    };
}
