// @specifyapp/sdk/src/builtInParsers/parsers-generation/to-react-native/toReactNativeHandler.ts
import { SDTFEngine, createSDTFEngine } from '@specifyapp/specify-design-token-format';
import { SpecifyError, specifyErrors } from '../../../errors/index.js';
import { FilesOutput } from '../../../parsersEngine/index.js';
import { generateCode } from '../../shared/to-javascript/generateCode.js';
import { tokensToAst } from '../../shared/to-javascript/tokensToAst.js';
import type { DeriveBuiltInParserHandlerFromDefinition } from '../../internals/createBuiltInParserDefinition.js';
import type { ToReactNativeParserDefinition } from './definition.js';
import { matcher } from './matcher.js';

export const toReactNativeHandler: DeriveBuiltInParserHandlerFromDefinition<
  ToReactNativeParserDefinition
> = async (previousDataBox, toolbox, parserOptions, outputOptions, _context) => {
  let sdtfEngine: SDTFEngine;

  switch (previousDataBox.type) {
    case 'SDTF': {
      sdtfEngine = createSDTFEngine(previousDataBox.graph, previousDataBox.metadata);
      break;
    }
    case 'SDTF Engine': {
      sdtfEngine = previousDataBox.engine;
      break;
    }
    default: {
      throw new SpecifyError({
        errorKey: specifyErrors.PARSERS_ENGINE_INVALID_PARSER_INPUT.errorKey,
        publicMessage: `${
          (previousDataBox as any).type
        } is not a valid input for the to-react-native parser.`,
      });
    }
  }

  if (outputOptions?.type !== 'file')
    throw new SpecifyError({
      errorKey: specifyErrors.PARSERS_ENGINE_INVALID_OUTPUT_TYPE.errorKey,
      publicMessage: `The output type ${outputOptions?.type} is not supported by the to-react-native parser.`,
    });

  let output: FilesOutput['files'] = [];

  const tokens = sdtfEngine.query.getAllTokenStates();

  if (tokens.length === 0) {
    toolbox.populateMessage({
      type: 'warning',
      content: `No design tokens found in the input`,
      errorKey: specifyErrors.PARSERS_ENGINE_PARSER_EXECUTION_FAILED.errorKey,
    });
  } else {
    const ast = tokensToAst(tokens, matcher);

    const exportStyle = parserOptions?.moduleExport ?? 'es6';

    output.push({
      path: outputOptions.filePath,
      content: {
        type: 'text',
        text: generateCode(ast, {
          isTypescript: parserOptions?.typescript ?? false,
          exportStyle,
          codeToPrepend:
            exportStyle === 'es6'
              ? "import { Easing } from 'react-native';"
              : "const { Easing } = require('react-native');",
        }),
      },
    });
  }

  toolbox.populateOutput({
    type: 'files',
    files: output,
  });

  return {
    type: 'SDTF Engine',
    engine: sdtfEngine,
  };
};
