// @specifyapp/specify-design-token-format/dist/engine/state/statefulValue/aliasing.js
/* ------------------------------------------
   Top Level Aliasing
--------------------------------------------- */
export class ResolvableTopLevelAlias {
    _kind = 'ResolvableTopLevelAlias';
    tokenState;
    constructor(params) {
        this.tokenState = params.tokenState;
    }
    get isFullyResolvable() {
        return this.tokenState.isFullyResolvable;
    }
}
export function matchIsResolvableTopLevelAlias(value) {
    return value instanceof ResolvableTopLevelAlias;
}
export class UnresolvableTopLevelAlias {
    _kind = 'UnresolvableTopLevelAlias';
    targetPath;
    constructor(params) {
        this.targetPath = params.targetPath;
    }
}
export function matchIsUnresolvableTopLevelAlias(value) {
    return value instanceof UnresolvableTopLevelAlias;
}
/* ------------------------------------------
   Mode Level Aliasing
--------------------------------------------- */
export class ResolvableModeLevelAlias {
    _kind = 'ResolvableModeLevelAlias';
    localMode;
    targetMode;
    tokenState;
    constructor(params) {
        this.localMode = params.localMode;
        this.targetMode = params.targetMode;
        this.tokenState = params.tokenState;
    }
    get isFullyResolvable() {
        return this.tokenState.modesResolvability[this.targetMode] === true;
    }
}
export function matchIsResolvableModeLevelAlias(value) {
    return value instanceof ResolvableModeLevelAlias;
}
export class UnresolvableModeLevelAlias {
    _kind = 'UnresolvableModeLevelAlias';
    localMode;
    targetMode;
    targetPath;
    constructor(params) {
        this.localMode = params.localMode;
        this.targetMode = params.targetMode;
        this.targetPath = params.targetPath;
    }
}
export function matchIsUnresolvableModeLevelAlias(value) {
    return value instanceof UnresolvableModeLevelAlias;
}
/* ------------------------------------------
   Value Level Aliasing
--------------------------------------------- */
export class ResolvableValueLevelAlias {
    _kind = 'ResolvableValueLevelAlias';
    localMode;
    valuePath;
    targetMode;
    tokenState;
    constructor(params) {
        this.localMode = params.localMode;
        this.valuePath = params.valuePath;
        this.targetMode = params.targetMode;
        this.tokenState = params.tokenState;
    }
    get isFullyResolvable() {
        return this.tokenState.modesResolvability[this.targetMode] === true;
    }
}
export function matchIsResolvableValueLevelAlias(value) {
    return value instanceof ResolvableValueLevelAlias;
}
export class UnresolvableValueLevelAlias {
    _kind = 'UnresolvableValueLevelAlias';
    localMode;
    valuePath;
    targetMode;
    targetPath;
    constructor(params) {
        this.localMode = params.localMode;
        this.valuePath = params.valuePath;
        this.targetMode = params.targetMode;
        this.targetPath = params.targetPath;
    }
}
export function matchIsUnresolvableValueLevelAlias(value) {
    return value instanceof UnresolvableValueLevelAlias;
}
