// @specifyapp/sdk/dist/fileSystem/fileSystem.js
import { promises as fs } from 'node:fs';
import * as nodePath from 'node:path';
/* v8 ignore start */
export function getDirectoryOfPath(path) {
    return nodePath.dirname(path);
}
export function joinPath(...paths) {
    return nodePath.join(...paths);
}
export function computeAbsolutePath(...relativePath) {
    return nodePath.resolve(process.cwd(), ...relativePath);
}
function pathStatsOrNull(path) {
    return fs.stat(path).catch(() => null);
}
export async function matchIsExistingDirectory(path) {
    const pathStats = await pathStatsOrNull(path);
    return !!(pathStats && pathStats.isDirectory());
}
export async function matchIsExistingFile(path) {
    const pathStats = await pathStatsOrNull(path);
    return !!(pathStats && pathStats.isFile());
}
export function isFilePath(path) {
    return nodePath.extname(path).length !== 0;
}
export async function generateEmptyFile(relativeFilePath) {
    const directoryPath = getDirectoryOfPath(relativeFilePath);
    if (!(await matchIsExistingDirectory(directoryPath))) {
        await fs.mkdir(directoryPath, {
            recursive: true,
        });
    }
    const absoluteFilePath = computeAbsolutePath(relativeFilePath);
    if (await matchIsExistingFile(absoluteFilePath)) {
        await fs.unlink(absoluteFilePath);
    }
    return absoluteFilePath;
}
/* v8 ignore stop */
export function makeFilePath(rootDirectoryPath, filePath) {
    let sanitizedRootDirectoryPath = rootDirectoryPath.trim();
    let sanitizedFilePath = filePath.trim();
    return nodePath.join(sanitizedRootDirectoryPath, sanitizedFilePath);
}
