// @specifyapp/specify-design-token-format/dist/engine/state/path/JSONPath.js
import { SDTF_PATH_SEPARATOR } from '../../../definitions/index.js';
import { arrayTrimStart } from '../../utils/arrayTrimStart.js';
export class JSONPath {
    #array;
    #string;
    constructor(path, string) {
        this.#array = path;
        this.#string = string;
    }
    clear() {
        this.#array = [];
        this.#string = '';
    }
    /**
     * We consider that the value is checked before reaching this class
     * So this function only do a basic join
     */
    #join() {
        return this.#array.join(SDTF_PATH_SEPARATOR);
    }
    get length() {
        return this.#array.length;
    }
    isRootOf(path) {
        return this.#array.every((pathItem, index) => {
            return path.at(index) === pathItem;
        });
    }
    isEqual(path) {
        return this.#string === path.#string;
    }
    isNotEqual(path) {
        return this.#string !== path.#string;
    }
    update(path, string) {
        this.#array = path;
        this.#string = string;
        return this;
    }
    pop() {
        const toReturn = this.#array.pop();
        this.#string = this.#join();
        return toReturn;
    }
    prepend(item) {
        this.#array.unshift(item);
        this.#string = this.#join();
        return this;
    }
    append(item) {
        this.#array.push(item);
        this.#string = this.#join();
        return this;
    }
    at(index) {
        return this.#array[index];
    }
    head() {
        return this.toArray()[0];
    }
    tail() {
        return this.toArray()[this.length - 1];
    }
    removeAt(i) {
        this.#array.splice(i, 1);
        this.#string = this.#join();
        return this;
    }
    replaceAt(i, value) {
        if (i >= this.#array.length) {
            return this;
        }
        this.#array[i] = value;
        this.#string = this.#join();
        return this;
    }
    removeLeft(until) {
        this.#array = this.#array.slice(until);
        this.#string = this.#join();
        return this;
    }
    removeRight(until) {
        this.#array = this.#array.slice(0, -1 * until);
        this.#string = this.#join();
        return this;
    }
    trimStartWith(path) {
        this.#array = arrayTrimStart(this.#array, path.toArray());
        this.#string = this.#join();
        return this;
    }
    concat(elements) {
        this.#array = this.#array.concat(elements);
        this.#string = this.#join();
        return this;
    }
    toString() {
        return this.#string;
    }
    // Needed for Vitest diff (called in JSON.stringify)
    toJSON() {
        return this.toArray();
    }
    toArray() {
        return this.#array;
    }
}
